---
title: "R Notebook"
output: html_notebook
---

```{r}
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(reshape2)
```
POSSIBLE OUTLIERS

GMCC4S10
GMCC3S7
GMCC4C6
GMCC5C5
GMCC5C4
GMCC5C3

9794IK
6238BB
control_b4_c6
control_b5_c5
control_b5_c4
control_b5_c3

Load metadata
```{r}
meta <- read.csv("data/metadata/meta_RNAseq.csv", row.names = 1) %>%
  # Batch as character
  mutate(Batch = as.character(Batch)) %>%
  # This sample was bad quality 
  dplyr::filter(!SampleID %in% c("GMCC1S1")) %>%
  # These samples were outliers amongst the RNAseq
  dplyr::filter(!SampleID %in% c("GMCC4S10", "GMCC3S7", "GMCC4C6", "GMCC5C5", "GMCC5C4", "GMCC5C3")) %>%
  # These samples were not used to treat the colonocytes
  dplyr::filter(!SampleName %in% c("1127XT", "1462QI", "3364PX", "3553XK", "3606IJ", "9423WC")) %>%
  # NOT SURE WHY THESE WERE REMOVED
  #dplyr::filter(!SampleID %in% c("GMCC1S6", "GMCC2S3", "GMCC4S7")) %>%
  # NOT SURE WHY THIS WAS REMOVED
  #dplyr::filter(!SampleName=="4342XH") %>%
  mutate(coculture = ifelse(grepl("control", SampleName), "control", "microbiome")) %>%
  # Replace 0s in urbanism with control
  mutate(urbanism = replace(urbanism, is.na(urbanism), "control")) %>%
  mutate(country = replace(country, is.na(country), "control")) %>%
  # Add one more column, urbanism2, which separates USA from non-USA urban
  mutate(urbanism2=ifelse(country=="USA", "USA_urban", urbanism)) %>%
  mutate(urbanism2 = replace(urbanism2, is.na(urbanism2), "control")) %>%
  mutate(rnames=SampleID) %>%
  column_to_rownames("rnames")

meta <- meta[order(row.names(meta)), ]
```
Update metadata with imputed values
```{r}
imputed_samples <- data.frame(
  SampleName = c("FMG110", "FMG111", "FMG112", "FMG28", "FMG65"),
  age = c(29,59,39,29,39),
  sex= c("F","M","F","M","M")
)
 
meta$age[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$age
meta$age <- as.numeric(meta$age)
meta$sex[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$sex
```


Import raw read counts
```{r}
raw_read_cts_filt <- read.csv("results/docs/raw_read_cts_filt.csv", row.names=1)
raw_read_cts_filt <- raw_read_cts_filt[, colnames(raw_read_cts_filt) %in% meta$SampleID]
raw_read_cts_filt <- raw_read_cts_filt[, order(colnames(raw_read_cts_filt))]
```

Run DESeq model
```{r}
dds <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt,
      colData = meta,
      design= ~Batch + urbanism))
```
```{r}
dds2 <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt,
      colData = meta,
      design= ~Batch + Total.Reads + coculture))
```

```{r}
vsd <- vst(dds2)
plotPCA(vsd, "Batch")
plotPCA(vsd, "coculture")
plotPCA(vsd, "urbanism")
assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$Batch)
plotPCA(vsd, "Batch")
plotPCA(vsd, "coculture")
plotPCA(vsd, "urbanism")
```
```{r}
pcaData <- plotPCA(vsd, intgroup=c("coculture", "urbanism", "Batch", "SampleName"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Batch, label=name)) +
  geom_text(size=2) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```


Results
```{r}
res_coculture <- results(dds2, contrast = c("coculture", "microbiome", "control")) %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")
  
res_uc <- results(dds, contrast = c("urbanism", "urban", "control")) %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")

res_rc <- results(dds, contrast = c("urbanism", "rural", "control")) %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")

res_ur <- results(dds, contrast = c("urbanism", "urban", "rural")) %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id")
```


Filter for significance
```{r}
sig_coculture <- res_coculture %>%
  dplyr::filter(padj<0.1) %>%
  #dplyr::filter(log2FoldChange > 0.1 | log2FoldChange < -0.1) %>%
  mutate(model="sig_coculture")
  
sig_uc <- res_uc %>%
  dplyr::filter(padj<0.1) %>%
  #dplyr::filter(log2FoldChange > 0.1 | log2FoldChange < -0.1) %>%
  mutate(model="sig_uc")

sig_uc_up <- sig_uc %>% dplyr::filter(log2FoldChange>0)
sig_uc_down <- sig_uc %>% dplyr::filter(log2FoldChange<0)

sig_rc <- res_rc %>%
  dplyr::filter(padj<0.1) %>%
  #dplyr::filter(log2FoldChange > 0.1 | log2FoldChange < -0.1) %>%
  mutate(model="sig_rc")

sig_rc_up <- sig_rc %>% dplyr::filter(log2FoldChange>0)
sig_rc_down <- sig_rc %>% dplyr::filter(log2FoldChange<0)

sig_ur <- res_ur %>%
  dplyr::filter(padj<0.1)
```
```{r}
library(EnhancedVolcano)
EnhancedVolcano(res_uc,
    lab = NA,
    x = 'log2FoldChange',
    y = 'pvalue',
    pCutoff = 0.1,
    FCcutoff = 0.1,
    pointSize = 3.0,
    labSize = 6.0)
```

L2FC scatterplot
```{r}
# Shared
sig_shared <- intersect(sig_uc$ensembl_gene_id, sig_rc$ensembl_gene_id)
# Unique
sig_uc_only <- sig_uc %>%
  dplyr::filter(!ensembl_gene_id %in% sig_shared)
sig_rc_only <- sig_rc %>%
  dplyr::filter(!ensembl_gene_id %in% sig_shared)

res_uc_lab <- res_uc %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, padj, pvalue) %>%
  mutate(neg_log10_pvalue = -log10(pvalue)) %>%
  mutate(neg_log10_padj = -log10(padj)) %>%
  rename_with(~paste0(., "_uc"), -1) 
res_rc_lab <- res_rc %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, padj, pvalue) %>%
  mutate(neg_log10_pvalue = -log10(pvalue)) %>%
  mutate(neg_log10_padj = -log10(padj)) %>%
  rename_with(~paste0(., "_rc"), -1)
  
res_uc_rc <- merge(
  res_uc_lab[,c("ensembl_gene_id", "log2FoldChange_uc", "padj_uc", "neg_log10_padj_uc", "pvalue_uc", "neg_log10_pvalue_uc")], 
  res_rc_lab[,c("ensembl_gene_id", "log2FoldChange_rc", "padj_rc", "neg_log10_padj_rc", "pvalue_rc", "neg_log10_pvalue_rc")], 
  by="ensembl_gene_id") %>%
  mutate(color_group = case_when(ensembl_gene_id %in% sig_shared ~ "Shared",
                                 ensembl_gene_id %in% sig_rc_only$ensembl_gene_id ~ "Rural only",
                                 ensembl_gene_id %in% sig_uc_only$ensembl_gene_id ~ "Urban only",
                                 !ensembl_gene_id %in% sig_shared ~ "z_Non-sig")) %>%
  mutate(sig = case_when(color_group != "z_Non-sig" ~ "Sig",
                         color_group == "z_Non-sig" ~ "z_Non-sig")) %>%
  mutate(abs_l2fc_uc = abs(log2FoldChange_uc)) %>%
  mutate(abs_l2fc_rc = abs(log2FoldChange_rc)) %>%
  arrange(desc(color_group))
```
Find most extreme genes belonging to uc or rc
```{r}
# Add an additional condition to exclude rows where padj or l2FC values are NA
res_uc_rc_rank <- res_uc_rc[complete.cases(res_uc_rc[, c("padj_uc", "padj_rc", "neg_log10_padj_uc", "neg_log10_padj_rc", "log2FoldChange_uc", "log2FoldChange_rc")]), ]
# Add a condition to exclude rows where padj < 0.01 in both uc and rc models
res_uc_rc_rank <- res_uc_rc_rank[(res_uc_rc_rank$padj_uc < 0.01 | res_uc_rc_rank$padj_rc < 0.01), ]

res_uc_rc_rank$neg_log10_padj_diff <- abs(res_uc_rc_rank$neg_log10_padj_uc - res_uc_rc_rank$neg_log10_padj_rc)
res_uc_rc_rank$l2FC_diff <- abs(res_uc_rc_rank$log2FoldChange_uc - res_uc_rc_rank$log2FoldChange_rc)

res_uc_rc_rank$neg_log10_padj_rank <- rank(-res_uc_rc_rank$neg_log10_padj_diff)
res_uc_rc_rank$l2FC_rank <- rank(-res_uc_rc_rank$l2FC_diff)

res_uc_rc_rank$rank_sum <- rowSums(res_uc_rc_rank[, c("neg_log10_padj_rank", "l2FC_rank")])

top_genes <- res_uc_rc_rank[order(res_uc_rc_rank$rank_sum), ][1:100, ] %>%
  mutate(most_extreme = "Most_Extreme: Urban v Control vs. Rural v Control")
```
```{r}
res_uc_rc_v2 <- res_uc_rc %>%
  left_join(top_genes[,c("ensembl_gene_id", "most_extreme")], by="ensembl_gene_id") %>%
  mutate(sig_extreme = case_when(most_extreme == "Most_Extreme: Urban v Control vs. Rural v Control" & sig=="Sig" ~ "Extreme",
                                 T ~ sig)) %>%
  arrange(desc(sig_extreme))
```

```{r}
ggplot(res_uc_rc, aes(x=neg_log10_pvalue_uc, y=neg_log10_pvalue_rc, color=color_group, fill=sig, shape=sig, size=sig)) +
  geom_point(alpha=0.5) +
  scale_shape_manual(values=c(19,3)) +
  scale_color_manual(values=c("#6FBD8B", "grey20", "#7B2BC1", "grey80")) + 
  scale_size_manual(values=c(1,1)) +
  geom_abline(linetype="dashed")+
  geom_smooth(data = subset(res_uc_rc_v2, sig == "Sig"), method = "lm", se = FALSE, size=0.5, color="indianred3") +
  xlab("neglog10 p: Industrialized vs control") +
  ylab("neglog10 p: Non-industrialized vs control") +
  #xlim(values=c(-0.75,0.75)) +
  #ylim(values=c(-0.75,0.75)) +
  xlim(values=c(0, 12.5)) +
  ylim(values=c(0, 12.5)) +
  theme_linedraw() + 
  coord_fixed() +
  theme(legend.position="none")

ggsave("results/figs/RNAseq-urbanism/pval_dist.png", width=4, height=3)

```
```{r}
most_sig <- rbind(res_uc, res_rc) %>%
  arrange(padj) %>%
  slice_head(n = 100) %>% #100
  distinct(ensembl_gene_id) %>%
  pull()
  
res_uc_lab2 <- res_uc %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, padj) %>%
  mutate(contrast = "uc")
res_rc_lab2 <- res_rc %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, padj) %>%
  mutate(contrast = "rc")

res_uc_rc_lab2 <-
  rbind(res_uc_lab2, res_rc_lab2) %>%
  left_join(res_uc_rc[,c("ensembl_gene_id", "sig", "color_group")], by="ensembl_gene_id") %>%
  dplyr::filter(sig=="Sig") %>%
  mutate(ind_sig = case_when(padj<0.005 ~ "Sig",
                             padj>0.005 ~ "NS")) %>%
  mutate(neg_log10_padj = -log10(padj)) %>%
  dplyr::filter(ensembl_gene_id %in% most_sig)

sorted_df <- res_uc_rc_lab2 %>%
  filter(contrast == "uc") %>%
  arrange(log2FoldChange)
  

res_uc_rc_lab2$ensembl_gene_id <- factor(res_uc_rc_lab2$ensembl_gene_id, levels=unique(sorted_df$ensembl_gene_id))
```
```{r}
ggplot(res_uc_rc_lab2, aes(x=log2FoldChange, y=ensembl_gene_id, fill=contrast, color=contrast, shape=ind_sig)) +
  geom_path(aes(group=ensembl_gene_id), color="grey", size=0.25) +
  geom_point(alpha=0.7) +
  scale_shape_manual(values=c(3,21)) +
  scale_color_manual(values=c("#457657", "#551D87")) +
  scale_fill_manual(values=c("#6FBD8B", "#7B2BC1")) +
  theme_linedraw() +
  theme(panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        #legend.position="none",
        axis.title.y = element_blank(),
        axis.text.y = element_blank()) +
  geom_vline(xintercept=0, color="grey")
ggsave("results/figs/RNAseq-urbanism/gene_l2fC_dotplot.png", width=3, height=7)
#ggsave("results/figs/RNAseq-urbanism/gene_l2fC_dotplot.png", width=3, height=4)
```


```{r}
library(lattice)
qqunif.plot<-function(pvalues, 
	should.thin=T, thin.obs.places=2, thin.exp.places=2, 
	xlab=expression(paste("Expected (",-log[10], " p-value)")),
	ylab=expression(paste("Observed (",-log[10], " p-value)")), 
	draw.conf=TRUE, conf.points=1000, conf.col="lightgray", conf.alpha=.05,
	already.transformed=F, pch=20, 
	#aspect="iso", 
	aspect=0.5,
	prepanel=prepanel.qqunif,
	par.settings=list(superpose.symbol=list(pch=pch)), ...) {
	
	
	#error checking
	if (length(pvalues)==0) stop("pvalue vector is empty, can't draw plot")
	if(!(class(pvalues)=="numeric" || 
		(class(pvalues)=="list" && all(sapply(pvalues, class)=="numeric"))))
		stop("pvalue vector is not numeric, can't draw plot")
	if (any(is.na(unlist(pvalues)))) stop("pvalue vector contains NA values, can't draw plot")
	if (already.transformed==FALSE) {
		if (any(unlist(pvalues)==0)) stop("pvalue vector contains zeros, can't draw plot")
	} else {
		if (any(unlist(pvalues)<0)) stop("-log10 pvalue vector contains negative values, can't draw plot")
	}
	
	
	grp<-NULL
	n<-1
	exp.x<-c()
	if(is.list(pvalues)) {
		nn<-sapply(pvalues, length)
		rs<-cumsum(nn)
		re<-rs-nn+1
		n<-min(nn)
		if (!is.null(names(pvalues))) {
			grp=factor(rep(names(pvalues), nn), levels=names(pvalues))
			names(pvalues)<-NULL
		} else {
			grp=factor(rep(1:length(pvalues), nn))
		}
		pvo<-pvalues
		pvalues<-numeric(sum(nn))
		exp.x<-numeric(sum(nn))
		for(i in 1:length(pvo)) {
			if (!already.transformed) {
				pvalues[rs[i]:re[i]] <- -log10(pvo[[i]])
				exp.x[rs[i]:re[i]] <- -log10((rank(pvo[[i]], ties.method="first")-.5)/nn[i])
			} else {
				pvalues[rs[i]:re[i]] <- pvo[[i]]
				exp.x[rs[i]:re[i]] <- -log10((nn[i]+1-rank(pvo[[i]], ties.method="first")-.5)/(nn[i]+1))
			}
		}
	} else {
		n <- length(pvalues)+1
		if (!already.transformed) {
			exp.x <- -log10((rank(pvalues, ties.method="first")-.5)/n)
			pvalues <- -log10(pvalues)
		} else {
			exp.x <- -log10((n-rank(pvalues, ties.method="first")-.5)/n)
		}
	}


#	#this is a helper function to draw the confidence interval
	panel.qqconf<-function(n, conf.points=1000, conf.col="gray", conf.alpha=.05, ...) {
		require(grid)
		conf.points = min(conf.points, n-1);
		mpts<-matrix(nrow=conf.points*2, ncol=2)
        	for(i in seq(from=1, to=conf.points)) {
            		mpts[i,1]<- -log10((i-.5)/n)
            		mpts[i,2]<- -log10(qbeta(1-conf.alpha/2, i, n-i))
            		mpts[conf.points*2+1-i,1]<- -log10((i-.5)/n)
            		mpts[conf.points*2+1-i,2]<- -log10(qbeta(conf.alpha/2, i, n-i))
        	}
        	grid.polygon(x=mpts[,1],y=mpts[,2], gp=gpar(fill=conf.col, lty=0), default.units="native")
    	}

	#reduce number of points to plot
	if (should.thin==T) {
		if (!is.null(grp)) {
			thin <- unique(data.frame(pvalues = round(pvalues, thin.obs.places),
				exp.x = round(exp.x, thin.exp.places),
				grp=grp))
			grp = thin$grp
		} else {
			thin <- unique(data.frame(pvalues = round(pvalues, thin.obs.places),
				exp.x = round(exp.x, thin.exp.places)))
		}
		pvalues <- thin$pvalues
		exp.x <- thin$exp.x
	}
	gc()
	
	prepanel.qqunif= function(x,y,...) {
		A = list()
		#A$xlim = range(x, y)*1.02
		A$xlim = c(0,5)
		A$xlim[1]=0
		#A$ylim = A$xlim
    A$ylim = c(0,15)
		return(A)
	}

	#draw the plot
	xyplot(pvalues~exp.x, groups=grp, xlab=xlab, ylab=ylab, aspect=1, #aspect
		prepanel=prepanel, scales=list(axs="i"), pch=pch,   
		panel = function(x, y, ...) {
			if (draw.conf) {
				panel.qqconf(n, conf.points=conf.points, 
					conf.col=conf.col, conf.alpha=conf.alpha)
			};
			panel.xyplot(x,y, ...);
			panel.abline(0,1);
		}, par.settings=par.settings, ...
	)
}
```

```{r}
select_v_all <- list("urban" = na.omit(res_uc$padj), "rural" = na.omit(res_rc$padj))
select_v_all_qq <- qqunif.plot(select_v_all, auto.key=list(corner=c(.95,.05)), col=c("#7B2BC1", "#6FBD8B"))
select_v_all_qq
```



How to derive residuals: https://support.bioconductor.org/p/60567/
```{r}
# Extract the residuals (batch and total.reads corrected counts)
fitted.common.scale = t(t(assays(dds)[["mu"]])/sizeFactors(dds))
residuals <- counts(dds, normalized=T) - fitted.common.scale

# function to filter by significant genes
filter_residuals <- function(residuals, sig_results) {
  filtered_residuals <- residuals[sig_results$ensembl_gene_id, , drop = FALSE]
  return(filtered_residuals)
}

# Use lapply to filter the "residuals" sub-element and retain "sig_results"
residuals_filt_uc <- filter_residuals(residuals, sig_uc) %>% 
    as.data.frame() %>% 
    rownames_to_column("ensembl_gene_id") %>%
  mutate(model="urban_v_ctrl")

residuals_filt_rc <- filter_residuals(residuals, sig_rc) %>% 
    as.data.frame() %>% 
    rownames_to_column("ensembl_gene_id") %>%
  mutate(model="rural_v_ctrl")

residuals_melt <- rbind(residuals_filt_rc, residuals_filt_uc) %>%
  pivot_longer(
    cols = -c("model", "ensembl_gene_id"),
    names_to = "SampleID",
    values_to = "residual_gene_count"
  ) %>%
  left_join(meta[,c("SampleID", "urbanism")], by="SampleID") %>%
  filter(!(model=="urban_v_ctrl" & urbanism=="rural")) %>%
  filter(!(model=="rural_v_ctrl" & urbanism=="urban")) 
```

Functional analysis
First annotate data with gene labels compatible with enrichment analyses. I like to have the uniprot gene symbol (hgnc_symbol) and a description just for my own readability.
```{r}
library(ensembldb)
library(biomaRt)
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl") #, host="jul2023.archive.ensembl.org"

# Define annotations
att_interest <- c("ensembl_gene_id", "description", "hgnc_symbol") #"ensembl_transcript_id", 

annotate_ensembl <- function(resdf){biomaRt::getBM(
  attributes = att_interest, 
  filters = "ensembl_gene_id",
  values = resdf$ensembl_gene_id, 
  mart = ensembl) %>%
  left_join(resdf, by="ensembl_gene_id")
}
```

Next define universe and gene set of interest. For my universe I used every gene I originally tested, and for sig I used the "top" significant genes (greatest l2FC + smallest p adjusted). Choosing the number of genes to test is a trade-off with multiple testing burden, but at least 100 genes will give you decent power. In this case my universe and sig are both dataframes, but character vectors are fine too.
Here I am experimenting with 3 different gene sets: "shared" (or "sig" here), "uc" and "rc". 
```{r}
# start with the genes that have the highest l2FCs and lowest padj values that are shared among the urban and rural

# dataframe of about 14,000 genes
uni <- unique(res_rc$ensembl_gene_id) %>%
  as.data.frame()
colnames(uni) <- "ensembl_gene_id"
uni <- annotate_ensembl(uni)

# dataframe of sig genes 
sig_top <- rbind(sig_uc, sig_rc) %>%
  dplyr::filter(ensembl_gene_id %in% sig_shared) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) %>%
  arrange(desc(abs_l2fc), padj) %>%
  annotate_ensembl() #%>%
  #slice(1:250)
#length(unique(sig_top$hgnc_symbol))

# another dataframe of sig genes 
uc_top <- sig_uc %>%
  #dplyr::filter(!ensembl_gene_id %in% sig_shared) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) %>%
  arrange(desc(abs_l2fc), padj) %>%
  annotate_ensembl() %>%
  dplyr::slice(1:400)
#length(unique(uc_top$hgnc_symbol))

# another dataframe of sig genes 
rc_top <- sig_rc %>%
  #dplyr::filter(!ensembl_gene_id %in% sig_shared) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) %>%
  arrange(desc(abs_l2fc), padj) %>%
  annotate_ensembl() %>%
  dplyr::slice(1:400)
#length(unique(rc_top$hgnc_symbol))
```
```{r}
sig_uc_up_anno <- annotate_ensembl(sig_uc_up) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) #%>%
  #dplyr::filter(abs_l2fc > 0.1)
sig_uc_down_anno <- annotate_ensembl(sig_uc_down) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) #%>%
  #dplyr::filter(abs_l2fc > 0.1)
sig_rc_up_anno <- annotate_ensembl(sig_rc_up) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) #%>%
  #dplyr::filter(abs_l2fc > 0.1)
sig_rc_down_anno <- annotate_ensembl(sig_rc_down) %>%
  mutate(abs_l2fc = abs(log2FoldChange)) #%>%
  #dplyr::filter(abs_l2fc > 0.1)

nrow(sig_uc_up_anno)
nrow(sig_uc_down_anno)
nrow(sig_rc_up_anno)
nrow(sig_rc_down_anno)
```

https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html
https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=CP
About these databases: https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp

Next choose the databases you want to test. KEGG and Reactome are the most relevant for my research.
Per Sambhawa's 2022 pub, I am only investigating a pathway if it contains at least 25 genes but less than 300 genes. 
```{r}
library(msigdbr)

#msigdbr_collections()

# download databases of interest
msigdb_h = msigdbr(species = "human", category = "H")
msigdb_cpPID = msigdbr(species = "human", category = "C2", subcategory = "CP:PID")
msigdb_cpKEGG = msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG")
msigdb_cpREACTOME = msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME")
msigdb_cpBIOCARTA = msigdbr(species = "human", category = "C2", subcategory = "CP:BIOCARTA")
msigdb_cgp = msigdbr(species = "human", category = "C2", subcategory = "CGP") %>%
  mutate(gs_name = paste("CGP", gs_name, sep = "_"))
msigdb_cgn = msigdbr(species = "human", category = "C4", subcategory = "CGN") %>%
  mutate(gs_name = paste("CGN", gs_name, sep = "_"))
msigdb_cm = msigdbr(species = "human", category = "C4", subcategory = "CM") %>%
  mutate(gs_name = paste("CM", gs_name, sep = "_"))
msigdb_IMMUNESIGDB = msigdbr(species = "human", category = "C7", subcategory = "IMMUNESIGDB") %>%
  mutate(gs_name = paste("IMMUNESIGDB", gs_name, sep = "_"))

# Experimenting with pasting these all together to reduce code burden
msigdb_cp <- rbind(msigdb_cpKEGG, msigdb_cpREACTOME) %>%
  rbind(msigdb_cpPID)

msigdb_all <- rbind(msigdb_cpKEGG, msigdb_cpREACTOME) %>%
  rbind(msigdb_cpPID) %>%
  rbind(msigdb_cpBIOCARTA) %>%
  rbind(msigdb_h) #%>%
  #rbind(msigdb_cgp) %>%
  #rbind(msigdb_cgn) %>%
  #rbind(msigdb_cm)  %>%
  #rbind(msigdb_IMMUNESIGDB)
  
# Summarize metrics from all pathways from all databases
msigdb_cp_summary <- msigdb_cp %>%
  group_by(gs_name) %>%
  summarize(GeneCount = n_distinct(gene_symbol)) %>%
  # sambhawas filtering cutoffs
  dplyr::filter(GeneCount > 25, GeneCount < 300)

msigdb_all_summary <- msigdb_all %>%
  group_by(gs_name) %>%
  summarize(GeneCount = n_distinct(gene_symbol)) %>%
  # sambhawas filtering cutoffs
  dplyr::filter(GeneCount > 25, GeneCount < 300)

# Narrow down pathways that meet our thresholds  
msigdb_cp_filt <- msigdb_cp %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpPID_filt <- msigdb_cpPID %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpKEGG_filt <- msigdb_cpKEGG %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpREACTOME_filt <- msigdb_cpREACTOME %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpBIOCARTA_filt <- msigdb_cpBIOCARTA %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)

msigdb_all_filt <- msigdb_all %>%
  dplyr::filter(gs_name %in% msigdb_all_summary$gs_name)
```
clusterProfiler for enrichment analysis. Once again I am trying 3 different GSEAs: "shared", "uc" and "rc"
```{r}
library(clusterProfiler)

# df of distinct gene symbols
# TERM2GENE is a data frame with first column of term ID and second column of corresponding mapped gene
msigdb_cp_t2g = msigdb_cp_filt %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
msigdb_all_t2g = msigdb_all_filt %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()

```
Not distinguishing up from down
```{r}

# Enrichment analysis step
# input to gene can be a df column or a list, same with universe
shared_enrich <- enricher(gene = unique(sig_top$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
shared_enrich_res <- shared_enrich@result %>%
  # I am only interested in a gene set if at least 3 of my sig genes belong, and if the FDR is < 0.1
  dplyr::filter(Count > 3, p.adjust < 0.1)

uc_enrich <- enricher(gene = unique(uc_top$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
uc_enrich_res <- uc_enrich@result %>% mutate(model="uc")
uc_enrich_sig <- uc_enrich_res %>% dplyr::filter(Count > 3, p.adjust < 0.1) 


rc_enrich <- enricher(gene = unique(rc_top$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
rc_enrich_res <- rc_enrich@result  %>% mutate(model="rc")
rc_enrich_sig <- rc_enrich_res %>% dplyr::filter(Count > 3, p.adjust < 0.1) 
```
Separate up from down
```{r}
uc_enrich_up <- enricher(gene = unique(sig_uc_up_anno$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
uc_enrich_up_res <- uc_enrich_up@result %>% mutate(model="uc") %>% mutate(dir="up")
uc_enrich_up_sig <- uc_enrich_up_res %>% dplyr::filter(Count > 3, p.adjust < 0.1)  

uc_enrich_down <- enricher(gene = unique(sig_uc_down_anno$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
uc_enrich_down_res <- uc_enrich_down@result %>% mutate(model="uc") %>% mutate(dir="down")
uc_enrich_down_sig <- uc_enrich_down_res %>% dplyr::filter(Count > 3, p.adjust < 0.1)  

rc_enrich_up <- enricher(gene = unique(sig_rc_up_anno$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
rc_enrich_up_res <- rc_enrich_up@result %>% mutate(model="rc") %>% mutate(dir="up")
rc_enrich_up_sig <- rc_enrich_up_res %>% dplyr::filter(Count > 3, p.adjust < 0.1)  

rc_enrich_down <- enricher(gene = unique(sig_rc_down_anno$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
rc_enrich_down_res <- rc_enrich_down@result %>% mutate(model="rc") %>% mutate(dir="down")
rc_enrich_down_sig <- rc_enrich_down_res %>% dplyr::filter(Count > 3, p.adjust < 0.1) 
```


Plot the results
```{r}
enriched_pathways <- #rbind(uc_enrich_res, rc_enrich_res) %>%
  rbind(uc_enrich_up_res, uc_enrich_down_res) %>%
  rbind(rc_enrich_up_res) %>%
  rbind(rc_enrich_down_res) %>%
  #dplyr::filter(ID %in% c(uc_enrich_sig$ID, rc_enrich_sig$ID)) %>%
  dplyr::filter(ID %in% c(uc_enrich_up_sig$ID, uc_enrich_down_sig$ID, rc_enrich_up_sig$ID, rc_enrich_down_sig$ID)) %>%
  mutate(db = case_when(
    grepl("HALLMARK", ID) ~ "Hallmark",
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    grepl("BIOCARTA", ID) ~ "Biocarta",
    grepl("CGP", ID) ~ "CGP",
    grepl("CGN", ID) ~ "CGN",
    grepl("CM", ID) ~ "CM",
    grepl("IMMUNESIGDB", ID) ~ "IMMUNESIGDB",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(ID_2 = gsub("^[^_]+_", "", ID)) %>%
  mutate(ID_2 = gsub("_", " ", ID_2)) 
  #mutate(ID_2 = paste0(toupper(substring(ID_2, 1, 1)), tolower(substring(ID_2, 2)))) %>%
  
```

```{r}
enriched_pathways_up <- enriched_pathways %>%
  dplyr::filter(dir=="up") %>%
  mutate(sig = case_when(
    model=="uc" & ID %in% uc_enrich_up_sig$ID ~ "sig_uc_up",
    model=="rc" & ID %in% rc_enrich_up_sig$ID ~ "sig_rc_up",
    model=="uc" & !ID %in% uc_enrich_up_sig$ID ~ "ns_uc_up",
    model=="rc" & !ID %in% rc_enrich_up_sig$ID ~ "ns_rc_up"
  )) %>%
  dplyr::filter(!sig=="ns_uc_up", !sig=="ns_rc_up") %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(neglog10_p.adjust = as.numeric(neglog10_p.adjust)) %>%
  mutate(GeneRatio_numeric = sapply(strsplit(as.character(GeneRatio), "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))) %>%
  mutate(sig_in = case_when(
    (ID %in% uc_enrich_up_sig$ID) & (ID %in% rc_enrich_up_sig$ID) ~ "both_up",
    (ID %in% uc_enrich_up_sig$ID) & (!ID %in% rc_enrich_up_sig$ID) ~ "urban_up_only",
    (ID %in% rc_enrich_up_sig$ID) & (!ID %in% uc_enrich_up_sig$ID) ~ "rural_up_only",
    TRUE ~ "Other"
  )) %>%
  arrange(GeneRatio_numeric)

enriched_pathways_down <- enriched_pathways %>%
  dplyr::filter(dir=="down") %>%
  mutate(sig = case_when(
    model=="uc" & ID %in% uc_enrich_down_sig$ID ~ "sig_uc_down",
    model=="rc" & ID %in% rc_enrich_down_sig$ID ~ "sig_rc_down",
    model=="uc" & !ID %in% uc_enrich_down_sig$ID ~ "ns_uc_down",
    model=="rc" & !ID %in% rc_enrich_down_sig$ID ~ "ns_rc_down",
  )) %>%
  dplyr::filter(!sig=="ns_uc_down", !sig=="ns_rc_down") %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(neglog10_p.adjust = as.numeric(neglog10_p.adjust)) %>%
  mutate(GeneRatio_numeric = sapply(strsplit(as.character(GeneRatio), "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))) %>%
  mutate(sig_in = case_when(
    (ID %in% uc_enrich_down_sig$ID) & (ID %in% rc_enrich_down_sig$ID) ~ "both_down",
    (ID %in% uc_enrich_down_sig$ID) & (!ID %in% rc_enrich_down_sig$ID) ~ "urban_down_only",
    (ID %in% rc_enrich_down_sig$ID) & (!ID %in% uc_enrich_down_sig$ID) ~ "rural_down_only",
    TRUE ~ "Other"
  )) %>%
  arrange(GeneRatio_numeric)

# I DON'T KNOW IF THIS IS THE BEST WAY TO DO THIS... 
# BUT TO REDUCE REDUNDANCY, PICK ID_2 WITH LARGEST PADJ
# IF SEVERAL HAVE SAME GENERATIO
enriched_pathways2_up <- enriched_pathways_up %>%
  group_by(sig_in, GeneRatio_numeric) %>%
  dplyr::slice(which.max(neglog10_p.adjust)) %>%
  ungroup() 

enriched_pathways_up$ID_2 <- factor(enriched_pathways_up$ID_2, levels=unique(enriched_pathways_up$ID_2))
enriched_pathways_up$sig_in <- factor(enriched_pathways_up$sig_in, levels=c("urban_up_only", "rural_up_only", "both_up"))

enriched_pathways2_up$ID_2 <- factor(enriched_pathways2_up$ID_2, levels=unique(enriched_pathways2_up$ID_2))
enriched_pathways2_up$sig_in <- factor(enriched_pathways2_up$sig_in, levels=c("urban_up_only", "rural_up_only", "both_up"))

enriched_pathways2_down <- enriched_pathways_down %>%
  group_by(sig_in, GeneRatio_numeric) %>%
  dplyr::slice(which.max(neglog10_p.adjust)) %>%
  ungroup() 

enriched_pathways_down$ID_2 <- factor(enriched_pathways_down$ID_2, levels=unique(enriched_pathways_down$ID_2))
enriched_pathways_down$sig_in <- factor(enriched_pathways_down$sig_in, levels=c("urban_down_only", "rural_down_only", "both_down"))

enriched_pathways2_down$ID_2 <- factor(enriched_pathways2_down$ID_2, levels=unique(enriched_pathways2_down$ID_2))
enriched_pathways2_down$sig_in <- factor(enriched_pathways2_down$sig_in, levels=c("urban_down_only", "rural_down_only", "both_down"))

ggplot(enriched_pathways2_up, aes(x=GeneRatio_numeric, y=ID_2, fill=model, color=model, size=neglog10_p.adjust)) + #, size=neglog10_p.adjust # shape=sig
  geom_point(alpha=0.75) +
  scale_size_continuous() +
  scale_color_manual(values = c("#6FBD8B", "#7B2BC1")) +
  scale_fill_manual(values = c("#6FBD8B", "#7B2BC1")) +
  scale_y_discrete(position = "right") +
  #scale_shape_manual(values = c(3,3,21,21)) +
  facet_grid(sig_in ~ ., scales="free", space="free") +
  theme_linedraw() +
  theme(text = element_text(size = 8),
        axis.text.y = element_text(hjust = 1))
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways_up.png", width=7.5, height=6)

ggplot(enriched_pathways2_down, aes(x=GeneRatio_numeric, y=ID_2, fill=model, color=model, size=neglog10_p.adjust)) + #, size=neglog10_p.adjust # shape=sig
  geom_point(alpha=0.75) +
  scale_size_continuous() +
  scale_color_manual(values = c("#6FBD8B", "#7B2BC1")) +
  scale_fill_manual(values = c("#6FBD8B", "#7B2BC1")) +
  scale_y_discrete(position = "right") +
  #scale_shape_manual(values = c(3,3,21,21)) +
  facet_grid(sig_in ~ ., scales="free", space="free") +
  theme_linedraw() +
  theme(text = element_text(size = 8),
        axis.text.y = element_text(hjust = 1))
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways_down.png", width=7.5, height=6)

#ggsave("results/figs/RNAseq-urbanism/enriched_pathways.png", width=11, height=6.5)
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways.png", width=5, height=3)
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways.png", width=7.5, height=6)
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways_updown.png", width=7.5, height=12)
```
Now plotting just the results of the "shared" model
```{r}
enriched_pathways_sharedgenesonly <- shared_enrich_res %>%
  mutate(db = case_when(
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(as.numeric(neglog10_p.adjust)) %>%
  arrange(neglog10_p.adjust) %>%
  mutate(value="value")
enriched_pathways_sharedgenesonly$ID <- factor(enriched_pathways_sharedgenesonly$ID, levels=unique(enriched_pathways_sharedgenesonly$ID))

ggplot(enriched_pathways_sharedgenesonly, aes(x=value, y=ID, size=neglog10_p.adjust)) +
  geom_point() +
  #scale_color_manual(values = c("#6FBD8B", "#7B2BC1")) +
  facet_grid(db ~ ., scales="free", space="free") +
  theme_linedraw()
#ggsave("results/figs/RNAseq-urbanism/enriched_pathways_sharedgenesonly.png", width=8, height=2)
```


GSEA analysis
```{r}
uc_gene_list = uc_top$abs_l2fc 
names(uc_gene_list) = uc_top$hgnc_symbol
uc_gene_list = sort(uc_gene_list, decreasing = TRUE) 
uc_gene_list = uc_gene_list[!duplicated(names(uc_gene_list))] 

rc_gene_list = rc_top$abs_l2fc 
names(rc_gene_list) = rc_top$hgnc_symbol
rc_gene_list = sort(rc_gene_list, decreasing = TRUE) 
rc_gene_list = rc_gene_list[!duplicated(names(rc_gene_list))] 


uc_gsea_res = GSEA(gene = uc_gene_list, TERM2GENE = msigdb_cp_t2g, pvalueCutoff = 0.1)
rc_gsea_res = GSEA(gene = rc_gene_list, TERM2GENE = msigdb_cp_t2g, pvalueCutoff = 0.1)
```


```{r}
rm(list = ls())
```