---
title: "R Notebook"
output: html_notebook
---

```{r}
library(ggplot2)
library(stringr)
library(pheatmap)
library(ensembldb)
library(biomaRt)
library(ggpubr)
library(tidyverse)
library(foreach)
library(DESeq2)
library(data.table)
library(ALDEx2)
```

--------------------------------------------------------------------------------
SETUP
--------------------------------------------------------------------------------
Load abundances of pathways
```{r}
#pathway_abundances <- read.csv("data/metaG-functional/pathabun_combined_20241021.tsv", header =T, sep = "\t")
pathway_abundances <- read.csv("data/metaG-functional/pathabun_combined.tsv", header =T, sep = "\t")
pathway_abundances2 <- pathway_abundances %>%
  # remove X from colnames
  rename_all(~gsub("^X", "", .)) %>%
  # remove "_Abundance" string from colnames
  rename_all(~str_remove(.x, "_Abundance")) %>%
  # rename first column
  rename_at(1, ~"Pathway") %>%
  # finally, tweak pathway names for compatibility with formula
  # tried simple quotes, which produced a few different errors
  # replace spaces with underscores 
  mutate(Pathway = str_replace_all(Pathway, " ", "_")) %>%
  # replace special characters with underscores 
  mutate(Pathway = str_replace_all(Pathway, "[^A-Za-z0-9_|]", "_")) %>%
  # add prefix "mc_" to Pathway names. prevents error if they start with a number
  mutate(Pathway = paste0("mc_", Pathway)) %>%
  # convert Pathway column to factor
  mutate(Pathway = as.factor(Pathway))
```
Read in pathway coverage
```{r}
#pathway_coverage <- read.csv("data/metaG-functional/pathcov_combined_20241021.tsv", header =T, sep = "\t")
pathway_coverage <- read.csv("data/metaG-functional/pathcov_combined.tsv", header =T, sep = "\t")
pathway_coverage2 <- pathway_coverage %>%
  # remove X from colnames
  rename_all(~gsub("^X", "", .)) %>%
  # remove "_Abundance" string from colnames
  rename_all(~str_remove(.x, "_Coverage")) %>%
  # rename first column
  rename_at(1, ~"Pathway") %>%
  # finally, tweak pathway names for compatibility with formula
  # tried simple quotes, which produced a few different errors
  # replace spaces with underscores 
  mutate(Pathway = str_replace_all(Pathway, " ", "_")) %>%
  # replace special characters with underscores 
  mutate(Pathway = str_replace_all(Pathway, "[^A-Za-z0-9_|]", "_")) %>%
  # add prefix "mc_" to Pathway names. prevents error if they start with a number
  mutate(Pathway = paste0("mc_", Pathway)) %>%
  # convert Pathway column to factor
  mutate(Pathway = as.factor(Pathway))
```
Gene families
```{r}
gene_fam <- read.csv("data/metaG-functional/genefam_combined_20241021.tsv", header =T, sep = "\t")
gene_fam2 <- gene_fam %>%
  # remove X from colnames
  rename_all(~gsub("^X", "", .)) %>%
  # remove "_Abundance" string from colnames
  rename_all(~str_remove(.x, "_Abundance.RPKs")) %>%
  # rename first column
  rename_at(1, ~"GeneFamily") %>%
  # finally, tweak pathway names for compatibility with formula
  # tried simple quotes, which produced a few different errors
  # replace spaces with underscores 
  mutate(GeneFamily = str_replace_all(GeneFamily, " ", "_")) %>%
  # replace special characters with underscores 
  mutate(GeneFamily = str_replace_all(GeneFamily, "[^A-Za-z0-9_|]", "_")) %>%
  # convert Pathway column to factor
  mutate(GeneFamily = as.factor(GeneFamily))
```
```{r}
head(gene_fam2)

gene_fam_r <- gene_fam2 %>% tibble::column_to_rownames("GeneFamily")
gene_fam_r[] <- lapply(gene_fam_r, function(x) x / sum(x))

head(gene_fam_r)
```




Import metadata
```{r}
meta <- read.csv("data/metadata/meta_RNAseq.csv", row.names = 1) 
# Identify missing samples
remove_from_abun <- setdiff(colnames(pathway_abundances2), meta$SampleName)
remove_from_meta <- setdiff(meta$SampleName, colnames(pathway_abundances2))
# Clean up metadata specific for this analysis
meta <- meta %>%
  # exclude controls
  dplyr::filter(!Luca_Lab_ID=="control") %>%
  # This sample was bad quality 
  dplyr::filter(!SampleID %in% c("GMCC1S1")) %>%
  # These samples were outliers amongst the RNAseq
  dplyr::filter(!SampleID %in% c("GMCC4S10", "GMCC3S7", "GMCC4C6", "GMCC5C5", "GMCC5C4", "GMCC5C3")) %>%
  # These samples were not used to treat the colonocytes
  dplyr::filter(!SampleName %in% c("1127XT", "1462QI", "3364PX", "3553XK", "3606IJ", "9423WC")) %>%
  # These samples did not have metagenomic sequencing data
  dplyr::filter(!SampleID %in% c("GMCC1S6", "GMCC2S3", "GMCC4S7")) %>%
  # exclude samples absent in pathway data
  dplyr::filter(!SampleName %in% remove_from_meta) %>%
  mutate(Batch = as.character(Batch)) %>%
  mutate(urbanism2=ifelse(country=="USA", "USA_urban", urbanism)) %>%
  mutate(SampleName2 = SampleName) %>%
  column_to_rownames("SampleName2") %>%
  dplyr::mutate(flag = case_when(
    country == "Ghana" ~ "data/metaG-QC/flags/ghana.svg",
    country == "Malaysia" ~ "data/metaG-QC/flags/malaysia.svg",
    country == "Nigeria" ~ "data/metaG-QC/flags/nigeria.svg",
    country == "Rwanda" ~ "data/metaG-QC/flags/rwanda.svg",
    country == "USA" ~ "data/metaG-QC/flags/usa.svg"))
meta <- meta[order(row.names(meta)), ]
```
Update metadata with imputed values
```{r}
imputed_samples <- data.frame(
  SampleName = c("FMG110", "FMG111", "FMG112", "FMG28", "FMG65"),
  age = c(29,59,39,29,39),
  sex= c("F","M","F","M","M")
)
 
meta$age[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$age
meta$age <- as.numeric(meta$age)
meta$sex[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$sex
```

```{r}
gene_fam_unmap <- gene_fam_r[c("UNMAPPED"),] %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("SampleName") %>%
  dplyr::mutate(MAPPED = 1-UNMAPPED) %>%
  left_join(meta[,c("urbanism", "SampleName")], by="SampleName") %>%
  dplyr::filter(!is.na(urbanism))

ggplot(gene_fam_unmap, aes(x=urbanism, y=MAPPED)) +
  geom_boxplot() +
  geom_point()

```

```{r}
gene_fam_unmap_summary <- gene_fam_unmap %>%
  group_by(urbanism) %>%
  summarize(median_MAPPED = median(MAPPED)) 
head(gene_fam_unmap_summary)
```




Now sort through this data to decide what to test. 

First let's filter for relevant pathways.
```{r}
pathway_abundances_filt1 <- pathway_abundances2 %>%
  # for now, filter at the pathway level, don't include microbe contributions
  dplyr::filter(!grepl("\\|", Pathway)) %>%
  # Remove unmapped and unintegrated
  dplyr::filter(!Pathway=="mc_UNMAPPED") %>%
  dplyr::filter(!Pathway=="mc_UNINTEGRATED") %>%
  dplyr::select(Pathway, meta$SampleName)
```
Remove pathways with a relative abundance of 0.01% in < 50% of rural OR < 50% of urban taxa.
```{r}
# Create list of the SampleIDs corresponding to different groups
urbanism_urban <- unlist(subset(meta$SampleName, meta$urbanism == "urban"))
urbanism_urban <- urbanism_urban[!(urbanism_urban %in% remove_from_meta)]
urbanism_rural <- unlist(subset(meta$SampleName, meta$urbanism == "rural"))
urbanism_rural <- urbanism_rural[!(urbanism_rural %in% remove_from_meta)]

pathway_abundances_filt1_r <- pathway_abundances_filt1 %>% tibble::column_to_rownames("Pathway")
pathway_abundances_filt1_r[] <- lapply(pathway_abundances_filt1_r, function(x) x / sum(x))

# Add columns to the counts matrix indicating the percentage of individuals with zero gene expression of each gene within a given group. 
pathway_abundances_r_filt2 <- pathway_abundances_filt1_r %>%
  dplyr::mutate(perc1en5_urbanism_urban = rowSums(across(all_of(urbanism_urban))<=1/10000)/length(urbanism_urban)) %>%
  dplyr::mutate(perc1en5_urbanism_rural = rowSums(across(all_of(urbanism_rural))<=1/10000)/length(urbanism_rural)) %>%
# Keep pathways which have abundance of 1e-4 values in less than or equal to 50% of urban group samples OR [cutoff value]% of rural group samples
  subset(perc1en5_urbanism_urban <= 0.50 | perc1en5_urbanism_rural <= 0.50) %>%
  dplyr::select(-(starts_with("perc1en5"))) %>%
  as.matrix()
  
# Apply filtering
keepPath = rownames(pathway_abundances_r_filt2)
pathway_abundances_filt2 = pathway_abundances_filt1 %>% dplyr::filter(Pathway %in% keepPath)

paste("Total number of pathways (filt cutoff = RA > 0.01%)")
paste("pre-filt: ", nrow(pathway_abundances_filt1))
paste0("post-filt: ", nrow(pathway_abundances_filt2))
```

Pare down pathway_coverage df for these pathways
```{r}
pathway_coverage_filt1 <- pathway_coverage2 %>%
  # for now, filter at the pathway level, don't include microbe contributions
  dplyr::filter(!grepl("\\|", Pathway)) %>%
  # Remove unmapped and unintegrated
  dplyr::filter(!Pathway=="UNMAPPED") %>%
  dplyr::filter(!Pathway=="UNINTEGRATED") %>%
  dplyr::select(Pathway %in% meta$SampleName) %>%
  dplyr::filter(Pathway %in% pathway_abundances_filt2$Pathway)
```


Visualize
```{r}
pathway_abundances_filt2_r <- pathway_abundances_filt2 %>% column_to_rownames("Pathway")
pathway_abundances_filt2_r[] <- lapply(pathway_abundances_filt2_r, function(x) x / sum(x))

pca_res <- prcomp(t(pathway_abundances_filt2_r), center = TRUE, scale. = TRUE)

# Extract the principal components
principal_components <- as.data.frame(pca_res$x[, 1:2]) %>%
  rownames_to_column("SampleName") %>%
  merge(meta[,c("SampleName", "urbanism", "flag", "country")], by="SampleName")
variance_explained <- 100 * pca_res$sdev^2 / sum(pca_res$sdev^2)

ggplot(principal_components, aes(x = PC1, y = PC2)) +
  geom_point() +
  labs(x = paste("PC1 [", round(variance_explained[1], 2), "%]", sep = ""),
       y = paste("PC2 [", round(variance_explained[2], 2), "%]", sep = "")) +
  geom_point(aes(fill=urbanism), shape=21, size=3, stroke = 0.75) +
  scale_fill_manual(values=c("#6FBD8B", "#7B2BC1"), name = "Population") + 
  theme_linedraw() +
  ggtitle("Pathway abundance of microbiomes (relative abundance)") +
  theme(
  axis.text = element_blank(),
  axis.ticks = element_blank()
  )
```
```{r}
library(ggimage)
library(rsvg)

ggplot(principal_components, aes(x = PC1, y = PC2, fill=country)) +
  geom_image(aes(image=flag), size=0.05) +
  #geom_point(aes(fill=country), shape=21, size=3, stroke = 0.75) +
  labs(x = paste("PC1 [", round(variance_explained[1], 2), "%]", sep = ""),
       y = paste("PC2 [", round(variance_explained[2], 2), "%]", sep = "")) +
  theme_linedraw() +
  ggtitle("Pathway abundance of microbiomes (relative abundance)") +
  theme(
  axis.text = element_blank(),
  axis.ticks = element_blank()
  )
```
Plot distribution of unmapped / unintegrated
```{r}
unmap_unint <- pathway_abundances_filt1_r %>% 
  #t() %>%
  rownames_to_column("Pathway") %>%
  dplyr::filter(str_detect(Pathway, "UNMAPPED")) %>%
  column_to_rownames("Pathway") %>%
  t() %>%
  as.data.frame() %>%
  dplyr::mutate(Mapped = (1 - (mc_UNMAPPED))) 
  #dplyr::select(1:2) %>%
  #dplyr::rename(UNMAPPED = 1) %>%
  #dplyr::rename(UNINTEGRATED = 2) %>%
  #rownames_to_column("SampleName") %>%
 #mutate(Mapped = (1 - (UNMAPPED + UNINTEGRATED))) %>%
#  pivot_longer(cols = -c("SampleName"),
#    names_to = "Category",
#    values_to = "Percentage"
#  ) #%>%
  #merge(meta[,c("SampleName", "urbanism")], by="SampleName")

ggplot(unmap_unint, aes(x=SampleName, y=Percentage, fill=Category)) +
  geom_bar(position="stack", stat="identity") #+
  #facet_grid(~urbanism, space="free", scales="free")

head(unmap_unint)
```

ALDEx2 for differential abundance testing. Negative effect size indicates the microbe is primarily found in rural, positive effect size means urban.

Explanation of output:
* rab.all - median clr value for all samples in the feature
* rab.win.NS - median clr value for the NS group of samples
* rab.win.S - median clr value for the S group of samples
* rab.X1_BNS.q50 - median expression value of features in sample X1_BNS if include.item.summary=TRUE
* dif.btw - median difference in clr values between S and NS groups
* dif.win - median of the largest difference in clr values within S and NS groups
* effect - median effect size: diff.btw / max(diff.win) for all instances
* overlap - proportion of effect size distribution that overlaps 0 (i.e. no effect)
∗ we.ep - Expected p-value of Welch’s t-test, a posterior predictive p-value
∗ we.eBH - Expected Benjamini-Hochberg corrected p-value of Welch’s t test
∗ wi.ep - Expected p-value of Wilcoxon rank test
∗ wi.eBH - Expected Benjamini-Hochberg corrected p-value of Wilcoxon test

I believe the order of operations within aldex is this:
1) Model data as dirichlet distribution
2) Monte Carlo
3) CLR transform
4) Differential abundance testing

Aldex2 for diet
Read in PC1 diet data from a later workbook


```{r}

aldex_glm_func <- function(df, var_interest, pval.holm) {

  # Extract the sample data from the phyloseq object and merge with df
  covariates <- meta %>%
    #data.frame() %>%
    merge(df[, c("SampleName", var_interest)], by = "SampleName") 

  # Remove samples with missing diet data
  pathway_abundance <- pathway_abundances_filt2 %>%
    column_to_rownames("Pathway") %>%
    #t() %>%
    #as.data.frame() %>%
    #rownames_to_column("SampleName") %>%
    dplyr::select(covariates$SampleName) %>%
    round() %>%
  as.matrix()

  # Select the variables of interest and other covariates
  covariates <- covariates %>%
    dplyr::select(all_of(var_interest), age, sex, country)

  # Create model matrix
  mm <- model.matrix(as.formula(paste("~", var_interest, "+ country + age + sex")), covariates)

  # Perform ALDEx2 analysis
  x.glm <- aldex.clr(pathway_abundance, mm, mc.samples = 128, denom = "all", verbose = TRUE)
  glm.test <- aldex.glm(x.glm, mm)
  glm.eff <- aldex.glm.effect(x.glm)

  # Combine results and add significance labels
  glm <- cbind(glm.eff[1], glm.test) %>%
    rename_with(~ str_replace_all(., ":", "_")) %>%
    mutate(sig = case_when({{pval.holm}} < 0.1 ~ "sig", TRUE ~ "ns")) %>%
    rownames_to_column("SampleName") %>%
    mutate(lab = case_when({{pval.holm}} < 0.1 ~ SampleName, TRUE ~ NA))

  # Filter for significant results
  glm_sig <- glm %>%
    dplyr::filter({{pval.holm}} < 0.1)

  return(glm_sig)
}


```

Get most abundant features (shooting for about 50?)
```{r}
pathway_abundances_filt2_mat_mostabun <- pathway_abundances_filt2_mat / rowSums(pathway_abundances_filt2_mat, na.rm = TRUE)
median_values <- apply(pathway_abundances_filt2_mat_mostabun, 1, median, na.rm = TRUE)
pathway_abundances_filt2_mat_mostabun <- cbind(pathway_abundances_filt2_mat_mostabun, median_abundance = median_values)

pathway_abundances_filt2_mat_mostabun_filt <- pathway_abundances_filt2_mat_mostabun %>%
  as.data.frame() %>%
  dplyr::filter(median_abundance > 0.01)

nrow(pathway_abundances_filt2_mat)
nrow(pathway_abundances_filt2_mat_mostabun_filt)

most_abundant_pathways <- unlist(row.names(pathway_abundances_filt2_mat_mostabun_filt))
write.csv(data.frame(taxa=most_abundant_pathways), file = "results/docs/pathways01.csv", row.names = FALSE)
```
Save pathway abundances
```{r}
write.csv(pathway_abundances_filt2, "results/docs/pathway_abundances_filt2.csv")
```


ALDeX2 with covariates (glm)
```{r}
pathway_abundances_filt2_mat <- pathway_abundances_filt2 %>%
  column_to_rownames("Pathway") %>%
  # round to integers
  round() %>%
  as.matrix()

covariates <- meta %>%
  dplyr::select(urbanism, country, age, sex)

mm <- model.matrix(~ urbanism, covariates)
x.glm <- aldex.clr(pathway_abundances_filt2_mat, mm, mc.samples=128, denom="all", verbose=T)
# Statistical results of coefficients
glm.test <- aldex.glm(x.glm, mm)
# Effect sizes
glm.eff<- aldex.glm.effect(x.glm)
glm.eff <- glm.eff[[1]]

# Filter for significance
glm.test_sig <- glm.test %>% dplyr::filter(`urbanismurban:pval.holm` < 0.1)
glm.eff_sig <- glm.eff[row.names(glm.eff) %in% row.names(glm.test_sig),]
```
```{r}
pathway_coverage_filt2_mat <- pathway_coverage_filt1 %>%
  column_to_rownames("Pathway") %>%
  # convert to percentage
 mutate(across(everything(), ~ .x * 100)) %>%
  # round to integers
  mutate(across(everything(), ~ round(.x, 0))) %>%
  as.matrix()

covariates <- meta %>%
  dplyr::select(urbanism, country, age, sex)

mm <- model.matrix(~ urbanism, covariates)
x.glm <- aldex.clr(pathway_coverage_filt2_mat, mm, mc.samples=128, denom="all", verbose=T)
# Statistical results of coefficients
glm.test <- aldex.glm(x.glm, mm)
# Effect sizes
glm.eff<- aldex.glm.effect(x.glm)
glm.eff <- glm.eff[[1]]

# Filter for significance
glm.test_sig <- glm.test %>% dplyr::filter(`urbanismurban:pval.holm` < 0.1)
glm.eff_sig <- glm.eff[row.names(glm.eff) %in% row.names(glm.test_sig),]
```
All results
```{r}
glm_urbanism <- cbind(glm.eff, glm.test) %>%
  dplyr::mutate(sig = case_when(`urbanismurban:pval.holm` < 0.1 ~ "sig",
                                 T ~ "ns")) %>%
  rownames_to_column("pathway") %>%
  dplyr::mutate(lab = case_when(sig=="sig" ~ pathway,
                                T ~ NA))
```

```{r}
library(ggrepel)

ggplot(glm_urbanism, aes(x=`urbanismurban:Est`, y=-log10(`urbanismurban:pval.holm`), color=sig, label=lab)) +
  geom_point() +
  scale_color_manual(values=c("grey", "tomato")) +
  geom_text_repel(size = 2,  # Adjust size as needed
                  nudge_y = 0.1) +  # Adjust nudge as needed
  theme_linedraw()
```
```{r}
pathway_coverage_filt1_plot <- pathway_coverage_filt1 %>%
  pivot_longer(cols = -Pathway, 
               names_to = "SampleName", 
               values_to = "coverage") %>%
  merge(meta[,c("SampleName", "urbanism")]) %>%
  mutate(coverage=as.numeric(coverage))

ggplot(
  subset(pathway_coverage_filt1_plot, Pathway=="mc_SER_GLYSYN_PWY__superpathway_of_L_serine_and_glycine_biosynthesis_I"),
  aes(x=urbanism, y=coverage)) +
  geom_boxplot() +
  geom_jitter(width=0.2)
```

mc_HSERMETANA_PWY__L_methionine_biosynthesis_III
mc_CENTFERM_PWY__pyruvate_fermentation_to_butanoate
mc_PWY_5676__acetyl_CoA_fermentation_to_butanoate_II
mc_FERMENTATION_PWY__mixed_acid_fermentation
mc_GLUCONEO_PWY__gluconeogenesis_I
mc_GLUDEG_I_PWY__GABA_shunt
mc_P108_PWY__pyruvate_fermentation_to_propanoate_I
mc_P41_PWY__pyruvate_fermentation_to_acetate_and__S__lactate_I
mc_PWY_5100__pyruvate_fermentation_to_acetate_and_lactate_II
mc_SER_GLYSYN_PWY__superpathway_of_L_serine_and_glycine_biosynthesis_I












Get CLR transformed counts
```{r}
aldex2_cts <- function(aldex_obj){
  x.counts <- getMonteCarloInstances(aldex_obj)
  # Calculate row means for each nested list
  row_means_list <- lapply(x.counts, function(i) {
    data.frame(rowMeans(i))
    })
  row_means_list <- lapply(names(row_means_list), function(list_name) {
    df <- row_means_list[[list_name]]
    colnames(df) <- list_name
    return(df)
    })
  # Convert the list of row means into a data frame and CLR transform
  row_means_df <- dplyr::bind_cols(row_means_list)
  return(row_means_df)
}

aldex2_clr_cts <- aldex2_cts(x.glm)
```
Export counts
```{r}
write.csv(aldex2_clr_cts, "results/docs/clr_filt_function_abundances.csv")
```







