---
title: "R Notebook"
output: html_notebook
---

```{r}
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(reshape2)
```
Load metadata
```{r}
richness <- read.csv("data/metadata/richness.csv", row.names = 1)
faith_pd <- read.csv("data/metadata/GMCC_covariates_all_batches_withFaith.csv") %>%
  dplyr::select(SampleID, Faith_PD)
meta <- read.csv("data/metadata/meta_RNAseq.csv", row.names = 1) %>%
 # merge(clr_genus_PC1PC2[,c("SampleName", "Axis1_genus", "Axis2_genus")], by="SampleName") %>%
  #dplyr::mutate(Axis1_genus = as.numeric(Axis1_genus)) %>%
  #mutate(Axis1_genus_scaled = scale(Axis1_genus)[, 1]) %>%
  #dplyr::mutate(Axis2_genus = as.numeric(Axis2_genus)) %>%
  #mutate(Axis2_genus_scaled = scale(Axis2_genus)[, 1]) %>%
  # Batch as character
  mutate(Batch = as.factor(Batch)) %>%
  # This sample was bad quality 
  dplyr::filter(!SampleID %in% c("GMCC1S1")) %>%
  # These samples were outliers amongst the RNAseq
  dplyr::filter(!SampleID %in% c("GMCC4S10", "GMCC3S7", "GMCC4C6", "GMCC5C5", "GMCC5C4", "GMCC5C3", "GMCC4S12")) %>% #NEW ADDITIONS AS OF NOV 2 2024 ARE GMCC4S12 AND GMCC3C1
  # These samples were not used to treat the colonocytes
  dplyr::filter(!SampleName %in% c("1127XT", "1462QI", "3364PX", "3553XK", "3606IJ", "9423WC")) %>%
  # NOT SURE WHY THESE WERE REMOVED
  #dplyr::filter(!SampleID %in% c("GMCC1S6", "GMCC2S3", "GMCC4S7")) %>%
  # NOT SURE WHY THIS WAS REMOVED
  #dplyr::filter(!SampleName=="4342XH") %>%
  mutate(coculture = ifelse(grepl("control", SampleName), "control", "microbiome")) %>%
  # Replace 0s in urbanism with control
  mutate(urbanism = replace(urbanism, is.na(urbanism), "control")) %>%
  mutate(country = replace(country, is.na(country), "control")) %>%
  mutate(age = replace(age, is.na(age), "control")) %>%
  mutate(sex = replace(sex, is.na(sex), "control")) %>%
  mutate(coculture_country = paste0(coculture, "_", country)) %>%
  # Add one more column, urbanism2, which separates USA from non-USA urban
  mutate(urbanism2=ifelse(country=="USA", "USA_urban", urbanism)) %>%
  mutate(urbanism2 = replace(urbanism2, is.na(urbanism2), "control")) %>%
  mutate(rnames=SampleID) %>%
  column_to_rownames("rnames") %>%
  mutate(Total.Reads_scaled = scale(Total.Reads)[, 1]) %>%
  #left_join(faith_pd, by="SampleID")
  left_join(richness[,c("SampleName", "Shannon")], by="SampleName") %>%
    dplyr::mutate(Shannon.y = as.numeric(Shannon.y)) 
  #dplyr::mutate(Chao1 = as.numeric(Chao1)) %>%
  #mutate(Chao1_scaled = scale(Chao1)[, 1]) 

meta <- meta[order(row.names(meta)), ]
```
Update metadata with imputed values
```{r}
imputed_samples <- data.frame(
  SampleName = c("FMG110", "FMG111", "FMG112", "FMG28", "FMG65"),
  age = c(29,59,39,29,39),
  sex= c("F","M","F","M","M")
)
 
meta$age[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$age
meta$age <- as.numeric(meta$age)
meta$sex[meta$SampleName %in% imputed_samples$SampleName] <- imputed_samples$sex
```
Import raw read counts
```{r}
raw_read_cts_filt <- read.csv("results/docs/raw_read_cts_filt.csv", row.names=1)
raw_read_cts_filt <- raw_read_cts_filt[, colnames(raw_read_cts_filt) %in% meta$SampleID]
raw_read_cts_filt <- raw_read_cts_filt[, order(colnames(raw_read_cts_filt))]
```

```{r}
meta_noctrl <- meta %>% dplyr::filter(!coculture=="control")  

raw_read_cts_filt_noctrl <- raw_read_cts_filt[,meta_noctrl$SampleID]
```
Only those with a value for shannon
```{r}
#meta_noshannonNA <- meta_noctrl %>% 
#  dplyr::filter(!is.na(Shannon.y)) 
#raw_read_cts_filt_noshannonNA <- raw_read_cts_filt_noctrl[,meta_noshannonNA$SampleID]

meta_noshannonNA <- meta_noctrl %>% 
  #dplyr::filter(!is.na(Faith_PD)) 
dplyr::filter(!is.na(Shannon.y)) 
raw_read_cts_filt_noshannonNA <- raw_read_cts_filt_noctrl[,meta_noshannonNA$SampleID]

```


```{r}
meta_ctrl <- meta %>%
  dplyr::filter(urbanism=="control") %>%
  dplyr::mutate(Shannon_group=0) #%>% 
  #mutate(age_scaled = scale(age)[, 1])

meta_noshannonNA_groups <- meta_noshannonNA %>% 
  #mutate(age_scaled = scale(age)[, 1]) %>% 
  arrange(Shannon.y) %>%  # Order by Shannon.y
  #arrange(Faith_PD) %>%  # Order by Shannon.y
  mutate(Shannon_group = ntile(Shannon.y, 2)) 
  #mutate(Shannon_group = ntile(Faith_PD, 2)) 

range(meta_noshannonNA$Shannon.y)
#range(meta_noshannonNA$Faith_PD)

meta_noshannonNA_groups_urbanism <- meta_noshannonNA %>% 
  #mutate(age_scaled = scale(age)[, 1]) %>% 
  arrange(Shannon.y) %>%  # Order by Shannon.y
  #arrange(Faith_PD) %>%  # Order by Shannon.y
  group_by(urbanism) %>%
  mutate(Shannon_group = ntile(Shannon.y, 2)) %>%
  #mutate(Shannon_group = ntile(Faith_PD, 2)) %>%
  ungroup() 


setdiff(colnames(meta_noshannonNA_groups_urbanism),colnames(meta_ctrl))

# all
meta_noshannonNA_withctrls <- meta_noshannonNA_groups %>% 
  rbind(meta_ctrl)
meta_noshannonNA_group1 <- meta_noshannonNA_groups %>% 
  dplyr::filter(Shannon_group==1) %>%
  rbind(meta_ctrl)
meta_noshannonNA_group2 <- meta_noshannonNA_groups %>% 
  dplyr::filter(Shannon_group==2) %>%
  rbind(meta_ctrl)

# urban only
meta_noshannonNA_urban_group1 <- meta_noshannonNA_groups_urbanism %>% 
  dplyr::filter(urbanism=="urban" & Shannon_group==1) %>%
  rbind(meta_ctrl)

meta_noshannonNA_urban_group2 <- meta_noshannonNA_groups_urbanism %>% 
  dplyr::filter(urbanism=="urban" & Shannon_group==2) %>%
  rbind(meta_ctrl)

# rural only
meta_noshannonNA_rural_group1 <- meta_noshannonNA_groups_urbanism %>% 
  dplyr::filter(urbanism=="rural" & Shannon_group==1) %>%
  rbind(meta_ctrl)

meta_noshannonNA_rural_group2 <- meta_noshannonNA_groups_urbanism %>% 
  dplyr::filter(urbanism=="rural" & Shannon_group==2) %>%
  rbind(meta_ctrl)


raw_read_cts_filt_noshannonNA_withctrls <- raw_read_cts_filt[,meta_noshannonNA_withctrls$SampleID]

raw_read_cts_filt_noshannonNA_urban_group1 <- raw_read_cts_filt[,meta_noshannonNA_urban_group1$SampleID]
raw_read_cts_filt_noshannonNA_urban_group2 <- raw_read_cts_filt[,meta_noshannonNA_urban_group2$SampleID]
raw_read_cts_filt_noshannonNA_rural_group1 <- raw_read_cts_filt[,meta_noshannonNA_rural_group1$SampleID]
raw_read_cts_filt_noshannonNA_rural_group2 <- raw_read_cts_filt[,meta_noshannonNA_rural_group2$SampleID]


print(paste("group1 urban",sum(meta_noshannonNA_urban_group1$urbanism == "urban", na.rm = TRUE)))
print(paste("group2 urban",sum(meta_noshannonNA_urban_group2$urbanism == "urban", na.rm = TRUE)))
print(paste("group1 rural",sum(meta_noshannonNA_rural_group1$urbanism == "rural", na.rm = TRUE)))
print(paste("group2 rural",sum(meta_noshannonNA_rural_group2$urbanism == "rural", na.rm = TRUE)))


```
```{r}
ggplot(meta_noshannonNA, aes(x=urbanism, y=Shannon.y, fill=urbanism)) +
 # ggplot(meta_noshannonNA, aes(x=urbanism, y=Faith_PD, fill=urbanism)) +
  geom_boxplot(alpha=0.8, color="black") +
  geom_jitter(width=0.2, size=1) +
  scale_fill_manual(values = c("#6FBD8B","#7B2BC1")) +
  scale_x_discrete(labels = c("Rural", "Urban")) +
  #labs(y = "Faith's Phylogenetic Diversity", x = NULL) +
  labs(y = "Shannon Diversity", x = NULL) +
  theme(legend.position = "none") +
  theme_linedraw() 
#ggsave("results/figs/RNAseq-shannon/ruralurban_faithpd.png", width=4, height=3)
ggsave("results/figs/RNAseq-shannon/ruralurban_shannon.png", width=4, height=3)


```


```{r}
ggplot(meta_noshannonNA_groups, aes(x=as.factor(Shannon_group), y=Shannon.y)) +
  #ggplot(meta_noshannonNA_groups, aes(x=as.factor(Shannon_group), y=Faith_PD)) +
  geom_boxplot() +
  geom_jitter(width = 0.2)

#wilcox.test(Faith_PD ~ urbanism, data = meta_noshannonNA)
wilcox.test(Shannon.y ~ urbanism, data = meta_noshannonNA)

ggplot(meta_noshannonNA_groups_urbanism, aes(x=factor(Shannon_group), y=Shannon.y)) +
  #ggplot(meta_noshannonNA_groups_urbanism, aes(x=factor(Shannon_group), y=Faith_PD)) +
  facet_wrap(~urbanism) +
  geom_boxplot() +
  geom_jitter(width = 0.2) 

ggplot(meta_noshannonNA_groups, aes(x=factor(Shannon_group), fill=urbanism)) +
  geom_bar() +
  scale_fill_manual(values = c("rural" = "#6FBD8B", "urban" = "#7B2BC1")) +
  theme_linedraw() +
  theme(legend.position="none")
ggsave("results/figs/RNAseq-shannon/shannon_dist_urbanrural_shannon.png", width=3, height=1.5)
#ggsave("results/figs/RNAseq-shannon/shannon_dist_urbanrural_faithpd.png", width=3, height=1.5)



# Compute density
dens <- density(meta_noshannonNA_groups$Shannon.y, na.rm = TRUE)
#dens <- density(as.numeric(meta_noshannonNA_groups$Faith_PD), na.rm = TRUE)
df_dens <- data.frame(x = dens$x, y = dens$y)

# Get median
med_val <- median(meta_noshannonNA_groups$Shannon.y, na.rm = TRUE)
#med_val <- median(as.numeric(meta_noshannonNA_groups$Faith_PD), na.rm = TRUE)

# Tag left and right of median
df_dens <- df_dens %>%
  mutate(side = ifelse(x <= med_val, "left", "right"))

# Plot
ggplot() +
  geom_area(data = df_dens %>% dplyr::filter(side == "left"),
            aes(x = x, y = y), fill = "#80ced7", alpha = 0.8) +
  geom_area(data = df_dens %>% dplyr::filter(side == "right"),
            aes(x = x, y = y), fill = "#007ea7", alpha = 0.8) +
  geom_rug(data = meta_noshannonNA_groups,
           aes(x = Shannon.y, color = urbanism),
           #aes(x = as.numeric(Faith_PD), color = urbanism),
           sides = "b", size = 0.5, length = unit(0.1, "npc")) +
  geom_vline(xintercept = med_val,
             linetype = "dashed", color = "black", size = 1) +
  scale_color_manual(values = c("rural" = "#6FBD8B", "urban" = "#7B2BC1")) +
  #ylim(values=c(-0.0025, 0.035)) +
  theme_linedraw() +
  #labs(x = "Shannon Diversity", y = "Density", color = "Urbanism") +
  theme(legend.position = "none")
#ggsave("results/figs/RNAseq-shannon/shannon_dist_faithpd.png", width=5, height=2)
ggsave("results/figs/RNAseq-shannon/shannon_dist_shannon.png", width=5, height=2)


# Step 1: Summarize counts and create labels
pie_data <- meta_noshannonNA_groups %>%
  count(Shannon_group, urbanism) %>%
  group_by(Shannon_group) %>%
  mutate(prop = n / sum(n),
         label = paste("n=", n, sep = ""))  # Create label with n=

# Step 2: Plot pie chart with labels
ggplot(pie_data, aes(x = "", y = prop, fill = urbanism)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar("y") +
  facet_wrap(~Shannon_group) +
  scale_fill_manual(values = c("rural" = "#6FBD8B", "urban" = "#7B2BC1")) +
  theme_void() +
  labs(fill = "Urbanism", title = "Urban vs Rural in Shannon Diversity Groups") +
  geom_text(aes(label = label), position = position_stack(vjust = 0.5), size = 5, color="white")  # Add labels to the slices
#ggsave("results/figs/RNAseq-shannon/shannon_pie_faithpd.png", width=4, height=2)
ggsave("results/figs/RNAseq-shannon/shannon_pie.png", width=4, height=2)

  
```
```{r}
meta_noshannonNA_groups_ordered <-  meta_noshannonNA_withctrls %>%
  dplyr::mutate(Shannon_group = as.factor(Shannon_group)) %>%
  dplyr::mutate(rnames = SampleID) %>%
  column_to_rownames("rnames")


# Reorder rows of metadata to match columns of count matrix
meta_noshannonNA_groups_ordered <- meta_noshannonNA_groups_ordered[colnames(raw_read_cts_filt_noshannonNA_withctrls), ]



dds_noshannonNA <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_withctrls,
      colData = meta_noshannonNA_groups_ordered,
      design= ~Batch + Total.Reads_scaled + Shannon_group))

```
```{r}
plotDispEsts(dds_noshannonNA)

```
```{r}


# Transform data with rlog or vst
rld <- vst(dds_noshannonNA, blind=TRUE)
plotPCA(rld, intgroup = c("Batch"))  # replace "condition" with your metadata column(s)

mat <- assay(rld)  # matrix of transformed counts (genes x samples)
pca <- prcomp(t(mat))  # transpose: samples in rows, genes in columns
loadings <- pca$rotation  # genes x PCs

top_genes <- sort(abs(loadings[,1]), decreasing=TRUE)
head(top_genes, 20)  # top 20 genes most strongly associated with PC1


# or
# rld <- vst(dds, blind=TRUE)

```








```{r}
resultsNames(dds_noshannonNA)
```


```{r}
res_shannon <- results(dds_noshannonNA, contrast = c("Shannon_group", "2", "1"))  %>%
  as.data.frame() %>%
  dplyr::mutate(sig = case_when(
    padj < 0.1  ~ "sig", 
    TRUE ~ "ns"
  )) %>%
  rownames_to_column("ensembl_gene_id")


sig_shannon <- res_shannon %>%
  dplyr::filter(sig=="sig")

sig_shannon_up <- sig_shannon %>%
  dplyr::filter(log2FoldChange>0)

sig_shannon_down <- sig_shannon %>%
  dplyr::filter(log2FoldChange<0)

#write.csv(sig_shannon_up, "results/docs/sig_highdivVlowdiv.csv")
#write.csv(sig_shannon_down, "results/docs/sig_lowdivVhighdiv.csv")
```

```{r}
res_group1 <- results(dds_noshannonNA, name = "Shannon_group_1_vs_0")  %>%
  as.data.frame() %>%
  dplyr::mutate(sig = case_when(
    padj < 0.1  ~ "sig", #& baseMean > 100
    TRUE ~ "ns"
  )) %>%
  rownames_to_column("ensembl_gene_id")


sig_group1 <- res_group1 %>%
  dplyr::filter(sig=="sig")

sig_up_group1 <- sig_group1 %>%
  dplyr::filter(log2FoldChange>0)

sig_down_group1 <- sig_group1 %>%
  dplyr::filter(log2FoldChange<0)

#write.csv(sig_up_group1, "results/docs/sig_lowdivVctrl_up.csv")
#write.csv(sig_down_group1, "results/docs/sig_lowdivVctrl_down.csv")


res_group2 <- results(dds_noshannonNA, name = "Shannon_group_2_vs_0")  %>%
  as.data.frame() %>%
  dplyr::mutate(sig = case_when(
    padj < 0.1  ~ "sig", #& baseMean > 100
    TRUE ~ "ns"
  )) %>%
  rownames_to_column("ensembl_gene_id")


sig_group2 <- res_group2 %>%
  dplyr::filter(sig=="sig")

sig_up_group2 <- sig_group2 %>%
  dplyr::filter(log2FoldChange>0)

sig_down_group2 <- sig_group2 %>%
  dplyr::filter(log2FoldChange<0)

#write.csv(sig_up_group2, "results/docs/sig_highdivVctrl_up.csv")
#write.csv(sig_down_group2, "results/docs/sig_highdivVctrl_down.csv")
```


Parse apart which groups have unique effects
```{r}
# assume res1 = group2_vs_group1, res2 = group1_vs_ctrl, res3 = group2_vs_ctrl
# all have columns: log2FoldChange, padj
library(dplyr)

combined <- sig_shannon %>%
  dplyr::select(ensembl_gene_id, lfc21=log2FoldChange, padj21=padj) %>%
  inner_join(res_group1 %>% dplyr::select(ensembl_gene_id, lfc1c=log2FoldChange, padj1c=padj), by="ensembl_gene_id") %>%
  inner_join(res_group2 %>% dplyr::select(ensembl_gene_id, lfc2c=log2FoldChange, padj2c=padj), by="ensembl_gene_id") %>%
  dplyr::mutate(class = case_when(
    padj1c < 0.1 & padj2c >= 0.1 ~ "LD differs from Control; HD unchanged",
    padj2c < 0.1 & padj1c >= 0.1 ~ "HD differs from Control; LD unchanged",
    padj1c < 0.1 & padj2c < 0.1 & sign(lfc1c)==sign(lfc2c) ~ "Both HD and LD differ from Control in the same direction",
    padj1c < 0.1 & padj2c < 0.1 & sign(lfc1c)!=sign(lfc2c) ~ "Both HD and LD differ from Control in the opposite directions",
    padj21 < 0.1 & padj1c >= 0.1 & padj2c >= 0.1 ~ "HD and LD differ from each other, but neither differs from Control",
    TRUE ~ "No effect"
  )) %>%
  dplyr::filter(!class=="No effect")

ggplot(combined, aes(x=class, fill=class)) +
  geom_bar(stat="count")

```



Create a bar plot that shows sig results
```{r}
sig_bar <- data_frame(labs=c("LD v. Control", "LD v. Control", "HD v. Control", "HD v. Control"), #, "LD v. HD", "LD v. HD"
                      #group=c("LD", "Control", "HD", "Control", "LD", "HD"),
                      n_genes= c(nrow(sig_up_group1), nrow(sig_down_group1), nrow(sig_up_group2), nrow(sig_down_group2))) #, nrow(sig_up), nrow(sig_down)

sig_bar$labs <- factor(sig_bar$labs, levels = c("LD v. Control", "HD v. Control")) #"LD v. HD", 

ggplot(sig_bar, aes(x = n_genes, y = labs, fill = labs)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = c("#80ced7",  "#007ea7")) + #"Control" = "#bfc0c0",
  labs(x = "Number of DEGs", y = NULL) +
  theme_linedraw() + 
  theme(legend.position="none") +
  coord_flip()
#ggsave("results/figs/RNAseq-shannon/DEGs_bar_faithpd.png", width=3, height=2)
ggsave("results/figs/RNAseq-shannon/DEGs_bar_shannon.png", width=3, height=2)
```



FUNCTIONAL ANALYSIS: OVER-REPRESENTATION
First annotate data with gene labels compatible with enrichment analyses. I like to have the uniprot gene symbol (hgnc_symbol) and a description just for my own readability.
```{r}
library(ensembldb)
library(biomaRt)


ensembl = biomaRt::useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl", host="jul2023.archive.ensembl.org") #, host="jul2023.archive.ensembl.org"

# Define annotations
att_interest <- c("ensembl_gene_id", "description", "hgnc_symbol", "entrezgene_id") #"ensembl_transcript_id", 

annotate_ensembl <- function(resdf){biomaRt::getBM(
  attributes = att_interest, 
  filters = "ensembl_gene_id",
  values = resdf$ensembl_gene_id, 
  mart = ensembl) %>%
  left_join(resdf, by="ensembl_gene_id")
}
```

Next define universe and gene set of interest. For my universe I used every gene I originally tested, and for sig I used the "top" significant genes (greatest l2FC + smallest p adjusted). Choosing the number of genes to test is a trade-off with multiple testing burden, but at least 100 genes will give you decent power. In this case my universe and sig are both dataframes, but character vectors are fine too.
```{r}
# start with the genes that have the highest l2FCs and lowest padj values that are shared among the urban and rural

# dataframe of about 14,000 genes
uni <- unique(res_shannon$ensembl_gene_id) %>%
  as.data.frame()
colnames(uni) <- "ensembl_gene_id"
uni <- annotate_ensembl(uni)

sig_shannon_anno <- sig_shannon %>%
  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "Enriched_HD",
            T ~ "Enriched_LD")) %>%
  annotate_ensembl()

sig_group1 <- sig_group1 %>%
  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "Upregulated",
            T ~ "Downregulated")) %>%
  annotate_ensembl()

sig_group2 <- sig_group2 %>%
  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "Upregulated",
            T ~ "Downregulated")) %>%
  annotate_ensembl()
```


Next choose the databases you want to test. KEGG and Reactome are the most relevant for my research.
Per Sambhawa's 2022 pub, I am only investigating a pathway if it contains at least 25 genes but less than 300 genes. 
```{r}
library(msigdbr)

#msigdbr_collections()

# download databases of interest
msigdb_h = msigdbr(species = "human", category = "H")
msigdb_cpPID = msigdbr(species = "human", category = "C2", subcategory = "CP:PID")
msigdb_cpKEGG = msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG_MEDICUS")
msigdb_cpREACTOME = msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME")
msigdb_cpBIOCARTA = msigdbr(species = "human", category = "C2", subcategory = "CP:BIOCARTA")
msigdb_GOBP = msigdbr(species = "human", category = "C5", subcategory = "GO:BP") %>%
  mutate(gs_name = paste("GOBP", gs_name, sep = "_"))
msigdb_GOBP = msigdbr(species = "human", category = "C5", subcategory = "GO:BP") %>%
  mutate(gs_name = paste("GOBP", gs_name, sep = "_"))

# Experimenting with pasting these all together to reduce code burden
msigdb_cp <- rbind(msigdb_cpKEGG, msigdb_cpREACTOME) %>%
  rbind(msigdb_cpPID)
  
  msigdb_all <- rbind(msigdb_cpKEGG, msigdb_cpREACTOME) %>%
  rbind(msigdb_cpPID) %>%
  rbind(msigdb_cpBIOCARTA) %>%
  rbind(msigdb_h)# %>%
  #rbind(msigdb_GOBP) #%>%
  #rbind(msigdb_cgn) %>%
  #rbind(msigdb_cm)  %>%
  #rbind(msigdb_IMMUNESIGDB)
  
# Summarize metrics from all pathways from all databases
msigdb_cp_summary <- msigdb_cp %>%
  group_by(gs_name) %>%
  dplyr::summarize(GeneCount = n_distinct(gene_symbol)) %>%
  # sambhawas filtering cutoffs
  dplyr::filter(GeneCount > 25, GeneCount < 300)

msigdb_GOBP_summary <- msigdb_GOBP %>%
  group_by(gs_name) %>%
  dplyr::summarize(GeneCount = n_distinct(gene_symbol)) %>%
  # sambhawas filtering cutoffs
  dplyr::filter(GeneCount > 25, GeneCount < 300)

msigdb_all_summary <- msigdb_all %>%
  group_by(gs_name) %>%
  dplyr::summarize(GeneCount = n_distinct(gene_symbol)) %>%
  # sambhawas filtering cutoffs
  dplyr::filter(GeneCount > 25, GeneCount < 300)

# Narrow down pathways that meet our thresholds  
msigdb_cp_filt <- msigdb_cp %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpPID_filt <- msigdb_cpPID %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpKEGG_filt <- msigdb_cpKEGG %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpREACTOME_filt <- msigdb_cpREACTOME %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_cpBIOCARTA_filt <- msigdb_cpBIOCARTA %>%
  dplyr::filter(gs_name %in% msigdb_cp_summary$gs_name)
msigdb_GOBP_filt <- msigdb_GOBP %>%
  dplyr::filter(gs_name %in% msigdb_GOBP_summary$gs_name)

msigdb_all_filt <- msigdb_all %>%
  dplyr::filter(gs_name %in% msigdb_all_summary$gs_name)
```
clusterProfiler for enrichment analysis.

```{r}
library(clusterProfiler)

# df of distinct gene symbols
# TERM2GENE is a data frame with first column of term ID and second column of corresponding mapped gene
msigdb_GOBP_t2g = msigdb_GOBP_filt %>% dplyr::distinct(gs_exact_source, gene_symbol) %>% as.data.frame()
msigdb_all_t2g = msigdb_all_filt %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()

```

```{r}
enrich_func <- function(sig, dir_in_quotes, enriched_in_quotes) {

  sig_dir <- sig %>% dplyr::filter(dir==dir_in_quotes) 
  enrich <- enricher(gene = unique(sig_dir$hgnc_symbol), TERM2GENE = msigdb_all_t2g, universe = uni$hgnc_symbol)
  enrich_res <- enrich@result %>% 
    mutate(enriched_in=enriched_in_quotes) %>%
    mutate(dir=dir_in_quotes)
  enrich_sig <- enrich_res %>% 
    dplyr::filter(Count > 2, p.adjust < 0.1) 
}

enrich_all_sig_down <- enrich_func(sig_shannon_anno, "Enriched_LD", "Low Diversity")
enrich_all_sig_up <- enrich_func(sig_shannon_anno, "Enriched_HD", "High Diversity")

enrich_group1_sig_up <- enrich_func(sig_group1, "Upregulated", "Low Diversity")
enrich_group1_sig_down <- enrich_func(sig_group1, "Downregulated", "Low Diversity")
enrich_group2_sig_up <- enrich_func(sig_group2, "Upregulated", "High Diversity")
enrich_group2_sig_down <- enrich_func(sig_group2, "Downregulated", "High Diversity")

unique_LD_up <- setdiff(rownames(enrich_group1_sig_up), rownames(enrich_group2_sig_up))
unique_LD_down <- setdiff(rownames(enrich_group1_sig_down), rownames(enrich_group2_sig_down))
unique_HD_up <- setdiff(rownames(enrich_group2_sig_up), rownames(enrich_group1_sig_up))
unique_HD_down <- setdiff(rownames(enrich_group2_sig_down), rownames(enrich_group1_sig_down))

# Indicate unique pathways
enrich_group1_sig_up$unique_pathway <- rownames(enrich_group1_sig_up) %in% unique_LD_up
enrich_group1_sig_down$unique_pathway <- rownames(enrich_group1_sig_down) %in% unique_LD_down
enrich_group2_sig_up$unique_pathway <- rownames(enrich_group2_sig_up) %in% unique_HD_up
enrich_group2_sig_down$unique_pathway <- rownames(enrich_group2_sig_down) %in% unique_HD_down

```

Also look at disease ontologies
```{r}
library(DOSE)
# Assume you have a vector of gene symbols called `gene_list`
# (convert to Entrez if needed using bitr from clusterProfiler)

# Perform enrichment
do_enrich_group1 <- enrichDO(gene          = enrich_group1_sig$entrezgene_id,
                      ont           = "DO",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable      = TRUE)
do_enrich_group1_sig <- do_enrich_group1@result %>%
  dplyr::filter(p.adjust < 0.1)

do_enrich_group2 <- enrichDO(gene          = enrich_group2_sig$entrezgene_id,
                      ont           = "DO",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable      = TRUE)
do_enrich_group2_sig <- do_enrich_group2@result %>%
  dplyr::filter(p.adjust < 0.1)

```
Protein coding genes identified in huang 2017
https://www.nature.com/articles/nature22969#MOESM1
```{r}
huang2017 <- unique(c("MMEL1", "MMEL1", "MMEL1", "IL23R", "IL23R", "IL23R", "PTPN22", "AP4B1", "ADAM15", "EFNA1", "ITLN1", "ITLN1", "FCGR2A", "FCGR2B", "IFIH1", "SP140", "ATG16L1", "MST1", "BSN", "MST1", "APEH", "BSN", "BSN", "MST1R", "SEMA3F", "COL7A1", "TMEM89", "CDHR4", "MST1", "IP6K2", "IFRD2", "BSN", "RBM6", "IP6K2", "AMT", "GPX1", "ARIH2", "CCDC71", "DAG1", "CCDC71", "MST1", "BANK1", "BANK1", "ERAP2", "ERAP2", "ERAP2", "ERAP2", "ERAP2", "IRGM", "TRAF3IP2", "REV3L", "SNAPC4", "CARD9", "SEC16A", "MAP3K8", "CUL2", "CD6", "NXPE1", "CXCR5", "LACC1", "GPR65", "SMAD3", "TNP2", "IL27", "SH2B1", "NOD2", "NOD2", "NOD2", "NOD2", "NOD2", "NOD2", "NOD2", "ZPBP2", "ZPBP2", "IKZF3", "TYK2", "KEAP1", "FUT2", "FUT2", "FUT2", "FUT2", "FUT2", "IRF5", "GSDMB"))

intersect(sig_up$hgnc_symbol, huang2017)
intersect(sig_down$hgnc_symbol, huang2017)
```




```{r}
test_enrichment_customgenelist <- function(sig_df, bg_df, disease_set) {
  # sig_up: data.frame or vector of gene symbols (e.g., sig_up$hgnc_symbol)
  # background: data.frame or vector of all possible gene symbols (e.g., uni$hgnc_symbol)
  # disease_set: vector of gene symbols (e.g., huang2017)
  
  # Convert to character vectors in case input is a data frame
  sig_genes <- unique(as.character(sig_df$hgnc_symbol))
  background_genes <- unique(as.character(bg_df$hgnc_symbol))
  disease_genes <- unique(as.character(disease_set))
  
  # Elements of the contingency table
  a <- length(intersect(sig_genes, disease_genes))
  b <- length(setdiff(sig_genes, disease_genes))
  c <- length(setdiff(disease_genes, sig_genes))
  d <- length(setdiff(background_genes, union(sig_genes, disease_genes)))
  
  # Create contingency table
  contingency_table <- matrix(c(a, b, c, d), nrow = 2,
                              dimnames = list(
                                Disease = c("In_GWASstudy", "Not_in_GWASstudy"),
                                SigUp = c("In_siggenes", "Not_in_siggenes")
                              ))
  
  # Print table and result
  print(contingency_table)
  fisher_res <- fisher.test(contingency_table, alternative = "greater")
  print(fisher_res)
  
  # Return the result as a list if needed programmatically
  invisible(list(
    table = contingency_table,
    fisher = fisher_res
  ))
}

test_enrichment_customgenelist(sig_up, uni, huang2017)
test_enrichment_customgenelist(sig_down, uni, huang2017)
```




Plot the results
```{#r}
enriched_pathways <- 
  rbind(enrich_up_res, enrich_down_res) %>%
  # only include those with count > 2
  dplyr::filter(Count > 2) %>%
  #dplyr::filter(ID %in% c(uc_enrich_sig$ID, rc_enrich_sig$ID)) %>%
  dplyr::filter(ID %in% c(enrich_down_sig$ID, enrich_down_sig$ID)) %>%
  mutate(db = case_when(
    grepl("HALLMARK", ID) ~ "Hallmark",
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    grepl("BIOCARTA", ID) ~ "Biocarta",
    grepl("CGP", ID) ~ "CGP",
    grepl("CGN", ID) ~ "CGN",
    grepl("CM", ID) ~ "CM",
    grepl("IMMUNESIGDB", ID) ~ "IMMUNESIGDB",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(ID_2 = gsub("^[^_]+_", "", ID)) %>%
  mutate(ID_2 = gsub("_", " ", ID_2)) %>%
  mutate(ID_3 = str_sub(ID_2, 1, 40)) 
  #mutate(ID_2 = paste0(toupper(substring(ID_2, 1, 1)), tolower(substring(ID_2, 2)))) %>%
  
```

```{r}
enrich_shannon <- rbind(enrich_all_sig_up, enrich_all_sig_down) %>%
  mutate(db = case_when(
    grepl("HALLMARK", ID) ~ "Hallmark",
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    grepl("BIOCARTA", ID) ~ "Biocarta",
    grepl("GOBP", ID) ~ "GOBP",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(ID_2 = gsub("^[^_]+_", "", ID)) %>%
  mutate(ID_2 = gsub("_", " ", ID_2)) %>%
  mutate(ID_3 = str_sub(ID_2, 1, 40)) %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(neglog10_p.adjust = as.numeric(neglog10_p.adjust)) %>%
  mutate(GeneRatio_numeric = sapply(strsplit(as.character(GeneRatio), "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))) %>%
  arrange(GeneRatio_numeric) %>%
  dplyr::mutate(unique_pathway=T)
  #dplyr::filter(db=="GOBP")

enrich_group1_sig <- rbind(enrich_group1_sig_up, enrich_group1_sig_down) %>%
  mutate(db = case_when(
    grepl("HALLMARK", ID) ~ "Hallmark",
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    grepl("BIOCARTA", ID) ~ "Biocarta",
    grepl("GOBP", ID) ~ "GOBP",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(ID_2 = gsub("^[^_]+_", "", ID)) %>%
  mutate(ID_2 = gsub("_", " ", ID_2)) %>%
  mutate(ID_3 = str_sub(ID_2, 1, 40)) %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(neglog10_p.adjust = as.numeric(neglog10_p.adjust)) %>%
  mutate(GeneRatio_numeric = sapply(strsplit(as.character(GeneRatio), "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))) %>%
  arrange(GeneRatio_numeric) #%>%
  #dplyr::filter(db=="GOBP")

enrich_group2_sig <- rbind(enrich_group2_sig_up, enrich_group2_sig_down) %>%
  mutate(db = case_when(
    grepl("HALLMARK", ID) ~ "Hallmark",
    grepl("REACTOME", ID) ~ "Reactome",
    grepl("KEGG", ID) ~ "KEGG",
    grepl("PID", ID) ~ "PID",
    grepl("BIOCARTA", ID) ~ "Biocarta",
    grepl("GOBP", ID) ~ "GOBP",
    TRUE ~ "Other"  # Add a default value for other cases
  )) %>%
  mutate(ID_2 = gsub("^[^_]+_", "", ID)) %>%
  mutate(ID_2 = gsub("_", " ", ID_2)) %>%
  mutate(ID_3 = str_sub(ID_2, 1, 40)) %>%
  mutate(neglog10_p.adjust = -log10(p.adjust)) %>%
  mutate(neglog10_p.adjust = as.numeric(neglog10_p.adjust)) %>%
  mutate(GeneRatio_numeric = sapply(strsplit(as.character(GeneRatio), "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))) %>%
  arrange(GeneRatio_numeric) #%>%
  #dplyr::filter(db=="GOBP")

enrich_shannon$ID_3 <- factor(enrich_shannon$ID_3, levels=unique(enrich_shannon$ID_3))
enrich_group1_sig$ID_3 <- factor(enrich_group1_sig$ID_3, levels=unique(enrich_group1_sig$ID_3))
enrich_group2_sig$ID_3 <- factor(enrich_group2_sig$ID_3, levels=unique(enrich_group2_sig$ID_3))

ggplot(enrich_shannon, aes(x=GeneRatio_numeric, y=ID_3, color=unique_pathway)) +
  geom_point() +
  facet_wrap(~dir, scales="free")
ggplot(enrich_group1_sig, aes(x=GeneRatio_numeric, y=ID_3, color=unique_pathway)) +
  geom_point() +
  facet_wrap(~dir, scales="free")
ggplot(enrich_group2_sig, aes(x=GeneRatio_numeric, y=ID_3, color=unique_pathway)) +
  geom_point()+
  facet_wrap(~dir, scales="free")
``` 
```{r}
dotplot_func <- function(df){
  df %>%
    filter(unique_pathway == TRUE) %>%
    group_by(dir) %>%
    arrange(desc(GeneRatio_numeric), desc(neglog10_p.adjust)) %>%
    ungroup() %>%
    distinct(GeneRatio_numeric, .keep_all = TRUE)
}

enrich_shannon_dotplot <- dotplot_func(enrich_shannon)
enrich_group1_sig_dotplot <- dotplot_func(enrich_group1_sig)
enrich_group2_sig_dotplot <- dotplot_func(enrich_group2_sig)

ggplot(enrich_shannon, aes(x=GeneRatio_numeric, y=ID_3, size=neglog10_p.adjust)) + #, size=neglog10_p.adjust # shape=sig
  geom_point(alpha=0.75, color="#80ced7", fill="#80ced7") +
  #scale_size_continuous(limits = c(1,15), breaks = c(1, 5, 10, 15, 20)) +
  scale_y_discrete(position = "right") +
  #scale_shape_manual(values = c(3,3,21,21)) +
  facet_grid(dir ~ ., scales="free", space="free") +
  theme_linedraw() +
  #xlim(values=c(0,0.20)) +
  theme(text = element_text(size = 8),
        axis.text.y = element_text(hjust = 1))

ggplot(enrich_group1_sig, aes(x=GeneRatio_numeric, y=ID_3, size=neglog10_p.adjust)) + #, size=neglog10_p.adjust # shape=sig
  geom_point(alpha=0.75, color="#80ced7", fill="#80ced7") +
  #scale_size_continuous(limits = c(1,15), breaks = c(1, 5, 10, 15, 20)) +
  scale_y_discrete(position = "right") +
  #scale_shape_manual(values = c(3,3,21,21)) +
  facet_grid(dir ~ ., scales="free", space="free") +
  theme_linedraw() +
  #xlim(values=c(0,0.20)) +
  theme(text = element_text(size = 8),
        axis.text.y = element_text(hjust = 1))

ggplot(enrich_group2_sig, aes(x=GeneRatio_numeric, y=ID_3, size=neglog10_p.adjust)) + #, size=neglog10_p.adjust # shape=sig
  geom_point(alpha=0.75, color="#007ea7", fill="#007ea7") +
  #scale_size_continuous(limits = c(1,15), breaks = c(1, 5, 10, 15, 20)) +
  scale_y_discrete(position = "right") +
  #scale_shape_manual(values = c(3,3,21,21)) +
  facet_grid(dir ~ ., scales="free", space="free") +
  theme_linedraw() +
  #xlim(values=c(0,0.20)) +
  theme(text = element_text(size = 8),
        axis.text.y = element_text(hjust = 1))
```

```{r}
library(ggfx)
library(ggbreak)
library(ggrepel)


ggplot(enrich_group1_sig, aes(x = GeneRatio_numeric, y = -log10(pvalue), fill=unique_pathway)) +
  scale_fill_manual(name = "Unique to LD", values=c("grey", "#80ced7")) +
  geom_vline(xintercept=0, color="grey", linetype="dashed") +
  geom_point(shape = 21, size = 3, color = "black", stroke = 0.5) +
  with_outer_glow(
     geom_text_repel(
  data = subset(enriched_pathways_group1, unique_pathway == TRUE),
  aes(label = ID_2), 
  color = "#80ced7", 
  segment.color = "black",
  box.padding = 0.3,
  size = 3,
  max.overlaps = Inf
) ,
  colour = "whitesmoke",   # glow color
  sigma = 1,          # blur size (adjust as needed)
  expand = 3          # how far glow extends (adjust as needed)
) +
  theme_linedraw() + 
  theme(legend.position = "none")
#ggsave("results/figs/RNAseq-shannon/enrichment_LDvCtrl_faithpd.svg", width=6, height=3.5)
ggsave("results/figs/RNAseq-shannon/enrichment_LDvCtrl_shannon.svg", width=6, height=3.5)

ggplot(enrich_group2_sig, aes(x = GeneRatio_numeric, y = -log10(pvalue), fill=unique_pathway)) +
  scale_fill_manual(name = "Unique to HD", values=c("grey", "#007ea7")) +
  geom_vline(xintercept=0, color="grey", linetype="dashed") +
  geom_point(shape = 21, size = 3, color = "black", stroke = 0.5) +
 with_outer_glow(
 geom_text_repel(
  data = subset(enriched_pathways_group2, unique_pathway == TRUE),
  aes(label = ID_2), 
  color = "#007ea7", 
  segment.color = "black",
  box.padding = 0.3,
  size = 3,
  max.overlaps = Inf
) ,
  colour = "whitesmoke",   # glow color
  sigma = 1,          # blur size (adjust as needed)
  expand = 3          # how far glow extends (adjust as needed)
) +
  theme_linedraw() + 
 # scale_y_break(c(10, 18), scales = 0.2) +
#  scale_y_continuous(breaks = seq(0, 20, by = 2), limits = c(0, 20)) +
  theme(legend.position = "none")
#ggsave("results/figs/RNAseq-shannon/enrichment_HDvCtrl_faithpd.svg", width=6, height=3.5)
ggsave("results/figs/RNAseq-shannon/enrichment_HDvCtrl_shannon.svg", width=6, height=3.5)
```


OBSERVED: The actual number of genes in your input gene set that are annotated to a given KEGG etc pathway. It is calculated from the GeneRatio column, which has the format x / y where:
x = number of genes in your input set that map to the pathway (the “observed” value)
y = total number of genes in your input set tested.

EXPECTED: The number of genes you would expect to see annotated to the pathway by chance, given the background distribution. It is calculated by first calculating the proportion of background genes annotated to the pathway:
bg_hits / bg_size
where:
bg_hits = total number of background genes annotated to the pathway
bg_size = total number of genes in the background set.

Then multiply this proportion by the number of genes tested (input_size), giving:
expected = input_size * (bg_hits / bg_size)

In words: If the pathway has, for example, 10% of background genes annotated to it, and you tested 200 genes, you’d expect ~20 genes annotated by chance (0.1 * 200). Comparing this to your observed value assesses enrichment.

```{r}
enriched_pathways3_func <- function(enriched_pathways2) {
  enriched_pathways3 <- enriched_pathways2 %>%
    mutate(
      observed = as.numeric(sub("/.*", "", GeneRatio)),
      input_size = as.numeric(sub(".*/", "", GeneRatio)),
      bg_hits = as.numeric(sub("/.*", "", BgRatio)),
      bg_size = as.numeric(sub(".*/", "", BgRatio)),
      expected = input_size * (bg_hits / bg_size)
    )
  
  x_range <- range(enriched_pathways3$expected, na.rm = TRUE)
  y_range <- range(enriched_pathways3$observed, na.rm = TRUE)
  diag_slope <- (y_range[2] - y_range[1]) / (x_range[2] - x_range[1])
  
  top_points2 <- enriched_pathways3 %>%
    # 1) Filter for unique_pathway == TRUE
    filter(unique_pathway == TRUE) %>%
    
    # 2) Arrange to prioritize highest GeneRatio_numeric then highest -log10(p.adjust)
    arrange(desc(GeneRatio_numeric), desc(neglog10_p.adjust)) %>%
    
    # 3) Keep distinct GeneRatio_numeric (i.e. only top per GeneRatio_numeric group)
    distinct(GeneRatio_numeric, .keep_all = TRUE) %>%
    
    # 4) Of these, keep top 5 overall by GeneRatio_numeric then -log10(p.adjust)
   # slice_max(order_by = GeneRatio_numeric, n = 15, with_ties = FALSE) %>%
    
    # Format labels
    mutate(ID_4 = str_replace(ID_3, "^(.{15,}?)\\s", "\\1\n")) %>%
    
    # Calculate label side
    mutate(
      diag_y_at_x = y_range[1] + diag_slope * (expected - x_range[1]),
      label_side = ifelse(observed > diag_y_at_x, "left", "right"),
      #nudge_x = ifelse(label_side == "left", (x_range/6)*1, x_range/6),
      #nudge_y = ifelse(label_side == "left", (y_range/5)*1, y_range/5)
      nudge_x = ifelse(label_side == "left", (x_range/7)*1, x_range/7),
      nudge_y = ifelse(label_side == "left", (y_range/7)*1, y_range/7)
    )
  
  return(list(enriched_pathways3, top_points2))
}

enriched_pathways3_shannon <- enriched_pathways3_func(enrich_shannon)
enriched_pathways3_group1 <- enriched_pathways3_func(enrich_group1_sig)
enriched_pathways3_group2 <- enriched_pathways3_func(enrich_group2_sig)
  

```
```{r}
library(ggfx)

labels_enrich_shannon <- enriched_pathways3_shannon[[2]] %>%
  group_by(dir) %>%
  slice_max(order_by = GeneRatio_numeric, n = 6, with_ties = FALSE) %>%
  ungroup()

ggplot(enriched_pathways3_shannon[[1]], aes(x = expected, y = observed, fill = dir)) +
  geom_point(shape = 24, size = 4, color = "black", stroke = 0.5) +
  scale_fill_manual(name = "Enriched in", values=c("#007ea7", "#80ced7")) +
  scale_color_manual(name = "Enriched in", values=c("#007ea7", "#80ced7")) +
  xlim(values=c(-0.02,25)) +
  ylim(values=c(-0.2,45)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", linewidth = 0.5) +
  with_outer_glow(
  geom_text_repel(
    data = labels_enrich_shannon,
    aes(label = ID_4, color=dir),  # text color by dir
    nudge_x = enriched_pathways3_shannon[[2]]$nudge_x,
    nudge_y = enriched_pathways3_shannon[[2]]$nudge_y,
    max.overlaps = Inf,
    size = 3
  ),
  colour = "whitesmoke",   # glow color
  sigma = 1,          # blur size (adjust as needed)
  expand = 3          # how far glow extends (adjust as needed)
) +
  theme_linedraw() +
  theme(legend.position = "none")

ggsave("results/figs/RNAseq-shannon/enrichment_HDvLD_shannon.png", width=6, height=3)

```
Save to file
```{r}
library(clipr)
write_clip(enriched_pathways3_shannon[[1]])
```



```{r}
ggplot(enriched_pathways3_group1[[1]], aes(x = expected, y = observed, fill = unique_pathway)) +
  geom_point(shape = 21, size = 4, color = "black", stroke = 0.5) +
  scale_fill_manual(name = "Unique to LD", values=c("grey", "#80ced7")) +
  scale_color_manual(name = "Unique to LD", values=c("grey", "#80ced7")) +
  xlim(values=c(-0.02,2.5)) +
  ylim(values=c(-0.2,20)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", linewidth = 0.5) +
  with_outer_glow(
  geom_text_repel(
    data = subset(enriched_pathways3_group1[[2]], unique_pathway==T),
    aes(label = ID_4),  # text color by dir
    nudge_x = enriched_pathways3_group1[[2]]$nudge_x,
    nudge_y = enriched_pathways3_group1[[2]]$nudge_y,
    max.overlaps = Inf,
    size = 3,
    color = "#80ced7"
  ),
  colour = "whitesmoke",   # glow color
  sigma = 1,          # blur size (adjust as needed)
  expand = 3          # how far glow extends (adjust as needed)
) +
  theme_linedraw() +
  theme(legend.position = "none")
#ggsave("results/figs/RNAseq-shannon/enrichment_LD_expobs_faithPD.png", width=6, height=3)
ggsave("results/figs/RNAseq-shannon/enrichment_LD_expobs_shannon.png", width=6, height=3)

ggplot(enriched_pathways3_group2[[1]], aes(x = expected, y = observed, fill = unique_pathway)) +
  geom_point(shape = 21, size = 4, color = "black", stroke = 0.5) +
  scale_fill_manual(name = "Unique to HD", values=c("grey", "#007ea7")) +
  scale_color_manual(name = "Unique to HD", values=c("grey", "#007ea7")) +
 # xlim(values=c(-0.2,6)) +
#  ylim(values=c(-0.2,22)) +
    xlim(values=c(-0.02,15)) +
  ylim(values=c(-0.2,30)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", linewidth = 0.5) +
  with_outer_glow(
  geom_text_repel(
    data =subset(enriched_pathways3_group2[[2]], unique_pathway==T & observed > 6),# enriched_pathways3_group1[[2]],
    aes(label = ID_4),  # text color by dir
    nudge_x = enriched_pathways3_group2[[2]]$nudge_x,
    nudge_y = enriched_pathways3_group2[[2]]$nudge_y,
    size = 3,
    color="#007ea7"
  ),
  colour = "whitesmoke",   # glow color
  sigma = 1,          # blur size (adjust as needed)
    expand = 3          # how far glow extends (adjust as needed)
) +
  theme_linedraw() +
  theme(legend.position = "none")


#ggsave("results/figs/RNAseq-shannon/enrichment_HD_expobs_faithPD.png", width=6, height=3)
ggsave("results/figs/RNAseq-shannon/enrichment_HD_expobs_shannon.png", width=6, height=3)
```



Genes to check from wastyk 2021
```{r}
wastyk21 <- c("NGF",
  "LIFR",
  "IL12B",
  "IL10",
  "CASP8",
  "FLICE",
  "MACH",
  "MCH5",
  "TGFB1",
  #"CD6",
  "CD5",
  "CCL8",
  "IL6",
  "CCL20",
  "LARC",
  "MIP3A",
  "ST38",
  "CKB4",
  "IL18",
  "VEGFA",
  "VGF",
  "VPF",
  "VEGF",
  "MMP10",
  "CCL13",
  "CCL4",
  "CXCL10",
  "CCL19"
  )


library(stringr)

sig_up_wastyk21 <- sig_up %>%
  filter(str_detect(hgnc_symbol, str_c(wastyk21, collapse = "|")))

sig_down_wastyk21 <- sig_down %>%
  filter(str_detect(hgnc_symbol, str_c(wastyk21, collapse = "|")))

```
```{r}

library(ggrepel)

genes_to_label <- c("NGFR", "TGFB1", "VGF", "IL6R")#, 
                   # "MMP10", "CXCL10", # from paper
                  #  "BST2", "IFI44L", "C3", "LCE1A")  

# Step 1: Combine sig_up and sig_down
#sig_combined <- bind_rows(sig_up, sig_down) %>%
#  distinct(ensembl_gene_id, hgnc_symbol)  # ensure uniqueness


# Step 2: Join with res_shannon
# Ensure interesting genes are plotted last (i.e., on top)
res_shannon_labeled <- res_shannon %>%
 # left_join(sig_combined, by = "ensembl_gene_id") %>%
  mutate(sig_direction = case_when(
    sig == "sig" & log2FoldChange < 0 ~ "left",
    sig == "sig" & log2FoldChange > 0 ~ "right",
    TRUE ~ "nonsig"
  )) %>%
  dplyr::filter(!padj==0)
 # mutate(label_flag = hgnc_symbol %in% genes_to_label) %>%
#  arrange(label_flag)  # FALSE (non-interesting) first, TRUE (interesting) last

ggplot(res_shannon_labeled, aes(x = log2FoldChange, y = -log10(pvalue), color = sig_direction)) +
  geom_point(alpha = 0.5) +
  scale_color_manual(values = c("left" = "#80ced7", "right" = "#007ea7", "nonsig" = "grey")) +
  geom_hline(yintercept = -log10(0.1)) +
  #geom_vline(xintercept = c(-1, 1)) +
  scale_fill_manual(values = c("left" = "#80ced7", "right" = "#007ea7", "nonsig" = "grey")) +
  theme_linedraw() +
  #xlim(values=c(-0.6,0.6)) +
 # geom_label_repel(
#    data = filter(res_shannon_labeled, hgnc_symbol %in% genes_to_label),
#    aes(label = hgnc_symbol, fill = sig_direction),
#    color = "white",           # text color
#    label.size = 0.2,          # border thickness
#    label.r = unit(0.15, "lines"), # small label corner radius
#    label.padding = unit(0.15, "lines"), # tighter padding
#      box.padding = 0.5,
#  point.padding = 0.2,
#  segment.color = "black",
#  segment.size = 0.7,
#  max.overlaps = Inf,
#  force = 1.5,
#    show.legend = FALSE
#  ) +
  theme(legend.position = "none")
ggsave("results/figs/RNAseq-shannon/genes_volcano_shannon.png", width=2.8, height=3)


# go read and find more specific genes
```

How to derive residuals: https://support.bioconductor.org/p/60567/
```{#r}
# Extract the residuals (batch and total.reads corrected counts)
fitted.values <- assays(dds_noshannonNA_group2)[["mu"]] #fitted values
fitted.common.scale = t(t(assays(dds_noshannonNA_group2)[["mu"]])/sizeFactors(dds_noshannonNA_group2)) # fitted values on common scale
residuals <- counts(dds_noshannonNA_group2, normalized=T)  # residuals
residuals.fitted <- counts(dds_noshannonNA_group2, normalized=T) - fitted.common.scale

# Prepare the VST matrix
vst_mat <- assay(vst(dds_noshannonNA_group2))

# Create a model matrix of the covariates you want to remove
design <- model.matrix(~ Batch, data = colData(dds_noshannonNA_group2))


# Remove the effects of all covariates using covariates argument
vst_corrected <- limma::removeBatchEffect(
  vst_mat,
  design = design
)



# function to filter by significant genes
filter_residuals <- function(residuals, list, results) {
  results_filt <- dplyr::filter(results, ensembl_gene_id %in% list)
  filtered_residuals <- residuals[results_filt$ensembl_gene_id, , drop = FALSE]
  return(filtered_residuals)
}

# Pick genes of interest
innate_immune_genes <- c(
  # review: https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2020.00965/full
  # also: https://www.nature.com/articles/s41423-021-00650-7
  

  "ENSG00000067057", #PFKP
  "ENSG00000067225", #PKM
  "ENSG00000074800", #ENO1
  "ENSG00000102144", #PGK1
  "ENSG00000111640", # GAPDH
  "ENSG00000149925", #ALDOA
  "ENSG00000105220", #GPI
  "ENSG00000172331" #BPGM
  
)


residuals_melt <- filter_residuals(vst_corrected, innate_immune_genes, res_group2) %>% 
    as.data.frame() %>% 
    rownames_to_column("ensembl_gene_id") %>%
  pivot_longer(
    cols = -c("ensembl_gene_id"),
    names_to = "SampleID",
    values_to = "residual_gene_count"
  ) %>%
  left_join(meta_noshannonNA_group2_ordered[,c("SampleID", "Shannon.y", "Shannon_group", "Batch")], by="SampleID") 
```


```{r}

# function to filter by significant genes
filter_residuals <- function(residuals, list, results) {
  results_filt <- dplyr::filter(results, ensembl_gene_id %in% list)
  filtered_residuals <- residuals[results_filt$ensembl_gene_id, , drop = FALSE]
  return(filtered_residuals)
}

# 1. Transform the data
vsd_func <- function(dds){
vsd <- vst(dds, blind = FALSE)
vsd_mat <- assay(vsd)  # This is used for batch correction
}

# 2. Remove batch effects
library(limma)

vsd_mat <- vsd_func(dds_noshannonNA)

vsd_corrected <- removeBatchEffect(vsd_mat,batch=colData(dds_noshannonNA)$Batch)
```


```{r}
# Pick genes of interest
genes_select <- c(
"ENSG00000075673" #ATP12A
#"ENSG00000151651", #ADAM8
#"ENSG00000100985", #MMP9
#"ENSG00000004799", #PDK4
#"ENSG00000115009", #CCL20
#"ENSG00000107201" #RIGI
)

gene_names <- c(
 "ENSG00000075673" ="ATP12A"
 #"ENSG00000151651" = "ADAM8",
#"ENSG00000100985" = "MMP9",
# "ENSG00000004799" = "PDK4",
#"ENSG00000115009" = "CCL20",
#"ENSG00000107201" = "RIGI"
)


residuals_melt <- filter_residuals(vsd_corrected, genes_select, res_shannon) %>% 
  as.data.frame() %>% 
  rownames_to_column("ensembl_gene_id") %>%
  pivot_longer(
    cols = -c("ensembl_gene_id"),
    names_to = "SampleID",
    values_to = "residual_gene_count"
  ) %>%
  mutate(gene_name = gene_names[ensembl_gene_id]) %>%
  left_join(
    meta_noshannonNA_groups_ordered[, c("SampleID", "Shannon.y", "Shannon_group", "urbanism", "coculture", "Batch", "country")],
    by = "SampleID"
  )

```



```{r}
ggplot(residuals_melt,
  aes(x=Shannon_group, y=residual_gene_count)) + #color= Batch,
  geom_boxplot(alpha=0.8, color="black") + 
  geom_jitter(width=0.2, size=1,  aes(color=country)) +
  scale_color_manual(values=c("black", "red", "gold", "green", "blue", "purple")) +
 # scale_fill_manual(values = c("#80ced7","#007ea7")) +
  facet_wrap(~gene_name, scales="free") +
  #scale_x_discrete(labels = c("1" = "Low \nDiversity", "2" = "High \nDiversity")) +
  theme_linedraw()
```
```{r}
ggplot(residuals_melt,
  aes(x=Shannon.y, y=residual_gene_count)) + #, color= Batch
  geom_point(shape=16, width = 0.1) + #alpha=0.4, 
  geom_smooth(method = "lm") +  
  facet_wrap(~gene_name, scales="free") +
  theme_linedraw()
```

```{r}
library(ggpubr)

boxplot_regression_func <- function(gene){

shannon_df <- residuals_melt %>% dplyr::filter(gene_name==gene)

p1 <- ggplot(shannon_df, aes(x=Shannon_group, y=residual_gene_count, fill=Shannon_group)) +
  geom_boxplot(alpha=0.8, color="black") +
  geom_jitter(width=0.2, size=1) +
  scale_fill_manual(values = c("#80ced7","#007ea7")) +
  scale_x_discrete(labels = c("1" = "Low \nDiversity", "2" = "High \nDiversity")) +
  labs(y = gene, x = NULL) +
  theme_linedraw() +
  theme(legend.position = "none")

p2 <- ggplot(shannon_df,
  aes(x=Shannon.y, y=residual_gene_count)) + #, color= Batch
  geom_point(shape=16, width = 0.1) + #alpha=0.4, 
  geom_smooth(method = "lm", color = "black") + 
  labs(y = gene, x = "Shannon diversity") +
  theme_linedraw()

ggarrange(p1,p2, nrow=1)
}

boxplot_regression_func("NGFR")
#ggsave("results/figs/RNAseq-shannon/ngfr.png", width=6, height=3)
boxplot_regression_func("COL1A2")
#ggsave("results/figs/RNAseq-shannon/col1a2.png", width=6, height=3)

boxplot_regression_func("MMP12")
ggsave("results/figs/RNAseq-shannon/mmp12.png", width=6, height=3)
```


```{r}
dds_noshannonNA_urban_group1 <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_urban_group1,
      colData = meta_noshannonNA_urban_group1,
      design= ~Batch + Total.Reads_scaled + coculture))

dds_noshannonNA_urban_group2 <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_urban_group2,
      colData = meta_noshannonNA_urban_group2,
      design= ~Batch + Total.Reads_scaled + coculture))

dds_noshannonNA_rural_group1 <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_rural_group1,
      colData = meta_noshannonNA_rural_group1,
      design= ~Batch + Total.Reads_scaled + coculture))

dds_noshannonNA_rural_group2 <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_rural_group2,
      colData = meta_noshannonNA_rural_group2,
      design= ~Batch + Total.Reads_scaled + coculture))

dds_noshannonNA_random <- DESeq(
    DESeqDataSetFromMatrix(
      countData = raw_read_cts_filt_noshannonNA_random,
      colData = meta_noshannonNA_random,
      design= ~Batch + Total.Reads_scaled + coculture))
```
```{r}
res_urban_group1 <- results(dds_noshannonNA_urban_group1, contrast=c("coculture", "control", "microbiome")) %>%
  as.data.frame()
res_urban_group2 <- results(dds_noshannonNA_urban_group2, contrast=c("coculture", "control", "microbiome")) %>%
  as.data.frame()
res_rural_group1 <- results(dds_noshannonNA_rural_group1, contrast=c("coculture", "control", "microbiome")) %>%
  as.data.frame()
res_rural_group2 <- results(dds_noshannonNA_rural_group2, contrast=c("coculture", "control", "microbiome")) %>%
  as.data.frame()
res_random <- results(dds_noshannonNA_random, contrast=c("coculture", "control", "microbiome")) %>%
  as.data.frame()
```

```{r}
sig_all_urbanism <- bind_rows(
  res_urban_group1 %>% filter(padj < 0.1) %>% mutate(pop="Industrialized", group = "Low Diversity"),
  res_urban_group2 %>% filter(padj < 0.1) %>% mutate(pop="Industrialized", group = "High Diversity"),
  res_rural_group1 %>% filter(padj < 0.1) %>% mutate(pop="Non-Industrialized", group = "Low Diversity"),
  res_rural_group2 %>% filter(padj < 0.1) %>% mutate(pop="Non-Industrialized", group = "High Diversity")
)

sig_all_urbanism$group <- factor(sig_all_urbanism$group, levels = c("Low Diversity", "High Diversity"))
levels(sig_all_urbanism$group) <- c("Low\nDiversity", "High\nDiversity")


degs_plot_urbanism <- ggplot(sig_all_urbanism, aes(x=group, fill=pop)) +
  geom_bar(stat="count") +
  scale_fill_manual(values=c("#7B2BC1", "#6FBD8B")) +
  facet_wrap(~pop) +
  labs(y = "Shannon diversity", x = NULL) +
  theme_linedraw() +
  theme(legend.position = "none") +
  ylab("Number of DEGs\nTreatment v. Control") 

degs_plot_urbanism

ggsave("results/figs/RNAseq-shannon/degs_plot_urbanism.png", width=6, height=3)
```

